---
title: "Massachusetts Data Processing for NFDS model"
output: html_notebook
editor_options: 
  chunk_output_type: console
---


```{r}
path_to_data <- "~/Documents/PhD_Project/Data/"
```

```{r}
### Reading in the Accession Codes, Population and the Sequence Clusters
library(readxl)
Samples_accCodes <- read_excel(paste(path_to_data, "Massachusetts_data_NickCroucher/SupplementaryDataPaper/Samples_accCodes.xlsx", sep = ""))
Mass_Samples_accCodes <- Samples_accCodes[Samples_accCodes$Population=="Massachusetts",]
Mass_Isolates <- Mass_Samples_accCodes$`Isolate Name`
Mass_Isolates_dict <- 1:length(Mass_Isolates)
names(Mass_Isolates_dict) <- Mass_Isolates

Isolates <- Samples_accCodes$`Isolate Name`
Isolates_dict <- 1:length(Isolates)
names(Isolates_dict) <- Isolates

Isolate_from_Mass_dict <- rep(0, length(Isolates))
names(Isolate_from_Mass_dict) <- Isolates
Isolate_from_Mass_dict[Mass_Isolates] <- 1
```

```{r}
### Reading in the Gene Cluster information
#gene_cluster <- read.table(paste(path_to_data, "v2_Massachusetts_data_NickCroucher/gCOG_sequences/CLS00005.out", sep = ""), quote="\"", comment.char="")[ c(TRUE,FALSE), ]

#function for reading cluster files
read_cls_file <- function(path_to_file){
  read.table(path_to_file, quote="\"", comment.char="")[ c(TRUE,FALSE), ]
}

#function for splitting the COG from the gene start
split_isolate_name <- function(text_line){
  name_vec <- strsplit(strsplit(text_line,">")[[1]][2],"_")
  paste(head(name_vec[[1]],-1), collapse = "_")
}
#isolates_gene_cluster <- unlist(lapply(gene_cluster, split_isolate_name))

# filter for Mass only
#Mass_isolates_gene_cluster <- isolates_gene_cluster[Isolate_from_Mass_dict[isolates_gene_cluster]==1]

# returns the Massachusetts-specific index of any Mass COG
return_isolate_index <- function(name){
  Mass_Isolates_dict[[name]]
}
#indices_gene_clusters <- unlist(lapply(Mass_isolates_gene_cluster, return_isolate_index))
```

```{r}
# list of input files
cls_files <- list.files(paste(path_to_data, "v2_Massachusetts_data_NickCroucher/gCOG_sequences", sep = ""), full.names = TRUE, recursive = FALSE)

# empty gene presence absence matrix
gene_presence_absence <- data.frame(matrix(0, nrow = length(cls_files)+1, ncol = length(Mass_Isolates)+1))
gene_presence_absence[1,-1] <- Mass_Isolates

# reads in all input files (takes a while, around 2000 files)
isolates_in_clusters <- (lapply(cls_files, read_cls_file))

# splits isolate names, changes values in gene presence absence matrix to 1 when genes are found
for (i in 1:length(isolates_in_clusters)) {
  isolates_gene_cluster <- unlist(lapply(isolates_in_clusters[[i]], split_isolate_name))
  Mass_isolates_gene_cluster <- isolates_gene_cluster[Isolate_from_Mass_dict[isolates_gene_cluster]==1]
  indices_gene_clusters <- unlist(lapply(Mass_isolates_gene_cluster, return_isolate_index))
  help_vector <- rep(0, length(Mass_Isolates))
  help_vector[indices_gene_clusters] <- 1
  gene_presence_absence[i+1,] <- c(strsplit(tail(strsplit(cls_files[i],"/")[[1]],1),"\\.")[[1]][1], help_vector)
}
```

```{r}
#saveRDS(gene_presence_absence, file = "gene_presence_absence.rds")
```


```{r}
# read in gene presence absence matrix
gene_presence_absence <- readRDS(file = "gene_presence_absence.rds")
```

```{r}
# create time-point specific gene presence absence matrices
Mass_year_dict <- c(rep(2001,length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2001,]$`Isolate Name`)), 
                    rep(2004,length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2004,]$`Isolate Name`)),
                    rep(2007,length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2007,]$`Isolate Name`)))
names(Mass_year_dict) <- c(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2001,]$`Isolate Name`, Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2004,]$`Isolate Name`, Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2007,]$`Isolate Name`)

# empty gene presence absence matrix
gene_presence_absence_2001 <- data.frame(matrix(0, nrow = length(cls_files)+1, ncol = length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2001,]$`Isolate Name`)+1))
gene_presence_absence_2001[1,-1] <- (Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2001,]$`Isolate Name`)
gene_presence_absence_2001[-1,1] <- gene_presence_absence[-1,1]
gene_presence_absence_2001[-1,-1] <- gene_presence_absence[-1,c(FALSE,Mass_year_dict[unlist(gene_presence_absence[1,-1])]==2001)]

gene_presence_absence_2004 <- data.frame(matrix(0, nrow = length(cls_files)+1, ncol = length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2004,]$`Isolate Name`)+1))
gene_presence_absence_2004[1,-1] <- (Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2004,]$`Isolate Name`)
gene_presence_absence_2004[-1,1] <- gene_presence_absence[-1,1]
gene_presence_absence_2004[-1,-1] <- gene_presence_absence[-1,c(FALSE,Mass_year_dict[unlist(gene_presence_absence[1,-1])]==2004)]

gene_presence_absence_2007 <- data.frame(matrix(0, nrow = length(cls_files)+1, ncol = length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2007,]$`Isolate Name`)+1))
gene_presence_absence_2007[1,-1] <- (Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2007,]$`Isolate Name`)
gene_presence_absence_2007[-1,1] <- gene_presence_absence[-1,1]
gene_presence_absence_2007[-1,-1] <- gene_presence_absence[-1,c(FALSE,Mass_year_dict[unlist(gene_presence_absence[1,-1])]==2007)]
```

```{r}
# Find intermediate frequency genes in 2001 data set
# compute gene frequencies
sum_as_int <- function(x){
  sum(as.integer(x))
}

gene_freq_2001 <- rep(0, nrow(gene_presence_absence_2001)-1)
gene_freq_2001 <- apply(gene_presence_absence_2001[-1,-1],1, sum_as_int)
gene_freq_2001 <- gene_freq_2001 / (length(gene_presence_absence_2001[1,-1]))

cog_filter <- as.integer(gene_freq_2001<=0.95 & gene_freq_2001>=0.05)

# intermediate gene presence absence matrices filtered by genes that are present at 5-95% frequency in 2001
intermed_gene_presence_absence_2001 <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2001,]$`Isolate Name`)+1))
intermed_gene_presence_absence_2001 <- gene_presence_absence_2001[c(1,which(cog_filter==1)+1),]

intermed_gene_presence_absence_2004 <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2004,]$`Isolate Name`)+1))
intermed_gene_presence_absence_2004 <- gene_presence_absence_2004[c(1,which(cog_filter==1)+1),]

intermed_gene_presence_absence_2007 <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2007,]$`Isolate Name`)+1))
intermed_gene_presence_absence_2007 <- gene_presence_absence_2007[c(1,which(cog_filter==1)+1),]

intermed_gene_presence_absence <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(Mass_Samples_accCodes$`Isolate Name`)+1))
intermed_gene_presence_absence <- gene_presence_absence[c(1,which(cog_filter==1)+1),]
```

```{r}
#save intermed gene presence absence matrices
#saveRDS(intermed_gene_presence_absence, file = "intermed_gene_presence_absence.rds")
#saveRDS(intermed_gene_presence_absence_2001, file = "intermed_gene_presence_absence_2001.rds")
#saveRDS(intermed_gene_presence_absence_2004, file = "intermed_gene_presence_absence_2004.rds")
#saveRDS(intermed_gene_presence_absence_2007, file = "intermed_gene_presence_absence_2007.rds")
```

```{r}
intermed_gene_presence_absence <- readRDS(file = "intermed_gene_presence_absence.rds")
intermed_gene_presence_absence_2001 <- readRDS(file = "intermed_gene_presence_absence_2001.rds")
intermed_gene_presence_absence_2004 <- readRDS(file = "intermed_gene_presence_absence_2004.rds")
intermed_gene_presence_absence_2007 <- readRDS(file = "intermed_gene_presence_absence_2007.rds")
```


```{r}
#Compute consensus genomes for sequence clusters
Mass_seq_clusters_dict <- Mass_Samples_accCodes$SequenceCluster
names(Mass_seq_clusters_dict) <- Mass_Samples_accCodes$`Isolate Name`

intermed_gene_presence_absence_consensus <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(unique(Mass_Samples_accCodes$SequenceCluster))+1))
intermed_gene_presence_absence_consensus[1,-1] <- paste("SeqCl_",unique(Mass_Samples_accCodes$SequenceCluster),sep = "")

intermed_gene_presence_absence_consensus_2001 <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(unique(Mass_Samples_accCodes$SequenceCluster))+1))
intermed_gene_presence_absence_consensus_2001[1,-1] <- paste("SeqCl_",unique(Mass_Samples_accCodes$SequenceCluster),sep = "")

intermed_gene_presence_absence_consensus_2004 <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(unique(Mass_Samples_accCodes$SequenceCluster))+1))
intermed_gene_presence_absence_consensus_2004[1,-1] <- paste("SeqCl_",unique(Mass_Samples_accCodes$SequenceCluster),sep = "")

intermed_gene_presence_absence_consensus_2007 <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(unique(Mass_Samples_accCodes$SequenceCluster))+1))
intermed_gene_presence_absence_consensus_2007[1,-1] <- paste("SeqCl_",unique(Mass_Samples_accCodes$SequenceCluster),sep = "")


cons_genomes <- function(x){
  as.double(ceiling(median(as.integer(x))))
}

for (i in unique(Mass_Samples_accCodes$SequenceCluster)) {
  intermed_gene_presence_absence_consensus[-1,i+2] <- apply(as.matrix(intermed_gene_presence_absence[-1,c(FALSE,Mass_seq_clusters_dict[unlist(intermed_gene_presence_absence[1,-1])]==i)]), 1, cons_genomes)
  intermed_gene_presence_absence_consensus_2001[-1,i+2] <- apply(as.matrix(intermed_gene_presence_absence_2001[-1,c(FALSE,Mass_seq_clusters_dict[unlist(intermed_gene_presence_absence_2001[1,-1])]==i)]), 1, cons_genomes)
    intermed_gene_presence_absence_consensus_2004[-1,i+2] <- apply(as.matrix(intermed_gene_presence_absence_2004[-1,c(FALSE,Mass_seq_clusters_dict[unlist(intermed_gene_presence_absence_2004[1,-1])]==i)]), 1, cons_genomes)
    intermed_gene_presence_absence_consensus_2007[-1,i+2] <- apply(as.matrix(intermed_gene_presence_absence_2007[-1,c(FALSE,Mass_seq_clusters_dict[unlist(intermed_gene_presence_absence_2007[1,-1])]==i)]), 1, cons_genomes)
}

intermed_gene_presence_absence_consensus[-1,1] <- intermed_gene_presence_absence[-1,1]
intermed_gene_presence_absence_consensus_2001[-1,1] <- intermed_gene_presence_absence[-1,1]
intermed_gene_presence_absence_consensus_2004[-1,1] <- intermed_gene_presence_absence[-1,1]
intermed_gene_presence_absence_consensus_2007[-1,1] <- intermed_gene_presence_absence[-1,1]
```

```{r}
#save intermed gene presence absence matrices
#saveRDS(intermed_gene_presence_absence_consensus, file = "intermed_gene_presence_absence_consensus.rds")
#saveRDS(intermed_gene_presence_absence_consensus_2001, file = "intermed_gene_presence_absence_consensus_2001.rds")
#saveRDS(intermed_gene_presence_absence_consensus_2004, file = "intermed_gene_presence_absence_consensus_2004.rds")
#saveRDS(intermed_gene_presence_absence_consensus_2007, file = "intermed_gene_presence_absence_consensus_2007.rds")
```

```{r}
# read in gene presence absence matrix
intermed_gene_presence_absence_consensus <- readRDS(file = "intermed_gene_presence_absence_consensus.rds")
intermed_gene_presence_absence_consensus_2001 <- readRDS(file = "intermed_gene_presence_absence_consensus_2001.rds")
intermed_gene_presence_absence_consensus_2004 <- readRDS(file = "intermed_gene_presence_absence_consensus_2004.rds")
intermed_gene_presence_absence_consensus_2007 <- readRDS(file = "intermed_gene_presence_absence_consensus_2007.rds")
```

```{r}
# calculate Vaccine Type consensus for clusters
mass_clusters <- length(unique(Mass_Samples_accCodes$SequenceCluster))
mass_VT <- rep(0, mass_clusters)

for (i in 1:mass_clusters){
  mass_VT[i] <- ceiling(median(as.integer(Mass_Samples_accCodes[Mass_Samples_accCodes$SequenceCluster == i-1,"Vaccine Type"]=="VT")))
}
# 1 means VT
# 0 means NVT
# if cluster is 50/50 it will count as VT
```
