---
title: "Massachusetts Data Processing for NFDS model"
output: html_notebook
editor_options: 
  chunk_output_type: console
---


```{r}
path_to_data <- "~/Documents/PhD_Project/Data/"
```

```{r}
### Reading in the Accession Codes, Population and the Sequence Clusters
library(readxl)
Samples_accCodes <- read_excel(paste(path_to_data, "Massachusetts_data_NickCroucher/SupplementaryDataPaper/Samples_accCodes.xlsx", sep = ""))
Mass_Samples_accCodes <- Samples_accCodes[Samples_accCodes$Population=="Massachusetts",]
Mass_Isolates <- Mass_Samples_accCodes$`Isolate Name`
Mass_Isolates_dict <- 1:length(Mass_Isolates)
names(Mass_Isolates_dict) <- Mass_Isolates

Isolates <- Samples_accCodes$`Isolate Name`
Isolates_dict <- 1:length(Isolates)
names(Isolates_dict) <- Isolates

Isolate_from_Mass_dict <- rep(0, length(Isolates))
names(Isolate_from_Mass_dict) <- Isolates
Isolate_from_Mass_dict[Mass_Isolates] <- 1
```

```{r}
#saveRDS(Mass_Samples_accCodes, file = "Mass_Samples_accCodes.rds")
```

```{r}
### Reading in the Gene Cluster information
#gene_cluster <- read.table(paste(path_to_data, "v2_Massachusetts_data_NickCroucher/gCOG_sequences/CLS00005.out", sep = ""), quote="\"", comment.char="")[ c(TRUE,FALSE), ]

#function for reading cluster files
read_cls_file <- function(path_to_file){
  read.table(path_to_file, quote="\"", comment.char="")[ c(TRUE,FALSE), ]
}

#function for splitting the COG from the gene start
split_isolate_name <- function(text_line){
  name_vec <- strsplit(strsplit(text_line,">")[[1]][2],"_")
  paste(head(name_vec[[1]],-1), collapse = "_")
}
#isolates_gene_cluster <- unlist(lapply(gene_cluster, split_isolate_name))

# filter for Mass only
#Mass_isolates_gene_cluster <- isolates_gene_cluster[Isolate_from_Mass_dict[isolates_gene_cluster]==1]

# returns the Massachusetts-specific index of any Mass COG
return_isolate_index <- function(name){
  Mass_Isolates_dict[[name]]
}
#indices_gene_clusters <- unlist(lapply(Mass_isolates_gene_cluster, return_isolate_index))
```

```{r}
# list of input files
cls_files <- list.files(paste(path_to_data, "v2_Massachusetts_data_NickCroucher/gCOG_sequences", sep = ""), full.names = TRUE, recursive = FALSE)

# empty gene presence absence matrix
gene_presence_absence <- data.frame(matrix(0, nrow = length(cls_files)+1, ncol = length(Mass_Isolates)+1))
gene_presence_absence[1,-1] <- Mass_Isolates

# reads in all input files (takes a while, around 2000 files)
isolates_in_clusters <- (lapply(cls_files, read_cls_file))

# splits isolate names, changes values in gene presence absence matrix to 1 when genes are found
for (i in 1:length(isolates_in_clusters)) {
  isolates_gene_cluster <- unlist(lapply(isolates_in_clusters[[i]], split_isolate_name))
  Mass_isolates_gene_cluster <- isolates_gene_cluster[Isolate_from_Mass_dict[isolates_gene_cluster]==1]
  indices_gene_clusters <- unlist(lapply(Mass_isolates_gene_cluster, return_isolate_index))
  help_vector <- rep(0, length(Mass_Isolates))
  help_vector[indices_gene_clusters] <- 1
  gene_presence_absence[i+1,] <- c(strsplit(tail(strsplit(cls_files[i],"/")[[1]],1),"\\.")[[1]][1], help_vector)
}
```

```{r}
#saveRDS(gene_presence_absence, file = "gene_presence_absence.rds")
```


```{r}
# read in gene presence absence matrix
gene_presence_absence <- readRDS(file = "gene_presence_absence.rds")
```

```{r}
# create time-point specific gene presence absence matrices
Mass_year_dict <- c(rep(2001,length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2001,]$`Isolate Name`)), 
                    rep(2004,length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2004,]$`Isolate Name`)),
                    rep(2007,length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2007,]$`Isolate Name`)))
names(Mass_year_dict) <- c(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2001,]$`Isolate Name`, Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2004,]$`Isolate Name`, Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2007,]$`Isolate Name`)

# empty gene presence absence matrix
gene_presence_absence_2001 <- data.frame(matrix(0, nrow = length(cls_files)+1, ncol = length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2001,]$`Isolate Name`)+1))
gene_presence_absence_2001[1,-1] <- (Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2001,]$`Isolate Name`)
gene_presence_absence_2001[-1,1] <- gene_presence_absence[-1,1]
gene_presence_absence_2001[-1,-1] <- gene_presence_absence[-1,c(FALSE,Mass_year_dict[unlist(gene_presence_absence[1,-1])]==2001)]

gene_presence_absence_2004 <- data.frame(matrix(0, nrow = length(cls_files)+1, ncol = length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2004,]$`Isolate Name`)+1))
gene_presence_absence_2004[1,-1] <- (Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2004,]$`Isolate Name`)
gene_presence_absence_2004[-1,1] <- gene_presence_absence[-1,1]
gene_presence_absence_2004[-1,-1] <- gene_presence_absence[-1,c(FALSE,Mass_year_dict[unlist(gene_presence_absence[1,-1])]==2004)]

gene_presence_absence_2007 <- data.frame(matrix(0, nrow = length(cls_files)+1, ncol = length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2007,]$`Isolate Name`)+1))
gene_presence_absence_2007[1,-1] <- (Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2007,]$`Isolate Name`)
gene_presence_absence_2007[-1,1] <- gene_presence_absence[-1,1]
gene_presence_absence_2007[-1,-1] <- gene_presence_absence[-1,c(FALSE,Mass_year_dict[unlist(gene_presence_absence[1,-1])]==2007)]
```

```{r}
# Find intermediate frequency genes in 2001 data set
# compute gene frequencies
sum_as_int <- function(x){
  sum(as.integer(x))
}

gene_freq_2001 <- rep(0, nrow(gene_presence_absence_2001)-1)
gene_freq_2001 <- apply(gene_presence_absence_2001[-1,-1],1, sum_as_int)
gene_freq_2001 <- gene_freq_2001 / (length(gene_presence_absence_2001[1,-1]))

cog_filter <- as.integer(gene_freq_2001<=0.95 & gene_freq_2001>=0.05)

# intermediate gene presence absence matrices filtered by genes that are present at 5-95% frequency in 2001
intermed_gene_presence_absence_2001 <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2001,]$`Isolate Name`)+1))
intermed_gene_presence_absence_2001 <- gene_presence_absence_2001[c(1,which(cog_filter==1)+1),]

intermed_gene_presence_absence_2004 <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2004,]$`Isolate Name`)+1))
intermed_gene_presence_absence_2004 <- gene_presence_absence_2004[c(1,which(cog_filter==1)+1),]

intermed_gene_presence_absence_2007 <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(Mass_Samples_accCodes[Mass_Samples_accCodes$`Year of Isolation`==2007,]$`Isolate Name`)+1))
intermed_gene_presence_absence_2007 <- gene_presence_absence_2007[c(1,which(cog_filter==1)+1),]

intermed_gene_presence_absence <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(Mass_Samples_accCodes$`Isolate Name`)+1))
intermed_gene_presence_absence <- gene_presence_absence[c(1,which(cog_filter==1)+1),]
```

```{r}
#save intermed gene presence absence matrices
#saveRDS(intermed_gene_presence_absence, file = "intermed_gene_presence_absence.rds")
#saveRDS(intermed_gene_presence_absence_2001, file = "intermed_gene_presence_absence_2001.rds")
#saveRDS(intermed_gene_presence_absence_2004, file = "intermed_gene_presence_absence_2004.rds")
#saveRDS(intermed_gene_presence_absence_2007, file = "intermed_gene_presence_absence_2007.rds")
```

```{r}
intermed_gene_presence_absence <- readRDS(file = "intermed_gene_presence_absence.rds")
intermed_gene_presence_absence_2001 <- readRDS(file = "intermed_gene_presence_absence_2001.rds")
intermed_gene_presence_absence_2004 <- readRDS(file = "intermed_gene_presence_absence_2004.rds")
intermed_gene_presence_absence_2007 <- readRDS(file = "intermed_gene_presence_absence_2007.rds")
```


```{r}
#Compute consensus genomes for sequence clusters
Mass_seq_clusters_dict <- Mass_Samples_accCodes$SequenceCluster
names(Mass_seq_clusters_dict) <- Mass_Samples_accCodes$`Isolate Name`

intermed_gene_presence_absence_consensus <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(unique(Mass_Samples_accCodes$SequenceCluster))+1))
intermed_gene_presence_absence_consensus[1,-1] <- paste("SeqCl_",unique(Mass_Samples_accCodes$SequenceCluster),sep = "")

intermed_gene_presence_absence_consensus_2001 <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(unique(Mass_Samples_accCodes$SequenceCluster))+1))
intermed_gene_presence_absence_consensus_2001[1,-1] <- paste("SeqCl_",unique(Mass_Samples_accCodes$SequenceCluster),sep = "")

intermed_gene_presence_absence_consensus_2004 <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(unique(Mass_Samples_accCodes$SequenceCluster))+1))
intermed_gene_presence_absence_consensus_2004[1,-1] <- paste("SeqCl_",unique(Mass_Samples_accCodes$SequenceCluster),sep = "")

intermed_gene_presence_absence_consensus_2007 <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(unique(Mass_Samples_accCodes$SequenceCluster))+1))
intermed_gene_presence_absence_consensus_2007[1,-1] <- paste("SeqCl_",unique(Mass_Samples_accCodes$SequenceCluster),sep = "")


cons_genomes <- function(x){
  as.double(ceiling(median(as.integer(x))))
}

for (i in unique(Mass_Samples_accCodes$SequenceCluster)) {
  intermed_gene_presence_absence_consensus[-1,i+2] <- apply(as.matrix(intermed_gene_presence_absence[-1,c(FALSE,Mass_seq_clusters_dict[unlist(intermed_gene_presence_absence[1,-1])]==i)]), 1, cons_genomes)
  intermed_gene_presence_absence_consensus_2001[-1,i+2] <- apply(as.matrix(intermed_gene_presence_absence_2001[-1,c(FALSE,Mass_seq_clusters_dict[unlist(intermed_gene_presence_absence_2001[1,-1])]==i)]), 1, cons_genomes)
    intermed_gene_presence_absence_consensus_2004[-1,i+2] <- apply(as.matrix(intermed_gene_presence_absence_2004[-1,c(FALSE,Mass_seq_clusters_dict[unlist(intermed_gene_presence_absence_2004[1,-1])]==i)]), 1, cons_genomes)
    intermed_gene_presence_absence_consensus_2007[-1,i+2] <- apply(as.matrix(intermed_gene_presence_absence_2007[-1,c(FALSE,Mass_seq_clusters_dict[unlist(intermed_gene_presence_absence_2007[1,-1])]==i)]), 1, cons_genomes)
}

intermed_gene_presence_absence_consensus[-1,1] <- intermed_gene_presence_absence[-1,1]
intermed_gene_presence_absence_consensus_2001[-1,1] <- intermed_gene_presence_absence[-1,1]
intermed_gene_presence_absence_consensus_2004[-1,1] <- intermed_gene_presence_absence[-1,1]
intermed_gene_presence_absence_consensus_2007[-1,1] <- intermed_gene_presence_absence[-1,1]
```

```{r}
#save intermed gene presence absence matrices
#saveRDS(intermed_gene_presence_absence_consensus, file = "intermed_gene_presence_absence_consensus.rds")
#saveRDS(intermed_gene_presence_absence_consensus_2001, file = "intermed_gene_presence_absence_consensus_2001.rds")
#saveRDS(intermed_gene_presence_absence_consensus_2004, file = "intermed_gene_presence_absence_consensus_2004.rds")
#saveRDS(intermed_gene_presence_absence_consensus_2007, file = "intermed_gene_presence_absence_consensus_2007.rds")
```

```{r}
# read in gene presence absence matrix
intermed_gene_presence_absence_consensus <- readRDS(file = "intermed_gene_presence_absence_consensus.rds")
intermed_gene_presence_absence_consensus_2001 <- readRDS(file = "intermed_gene_presence_absence_consensus_2001.rds")
intermed_gene_presence_absence_consensus_2004 <- readRDS(file = "intermed_gene_presence_absence_consensus_2004.rds")
intermed_gene_presence_absence_consensus_2007 <- readRDS(file = "intermed_gene_presence_absence_consensus_2007.rds")
```

```{r}
# calculate Vaccine Type consensus for clusters
mass_clusters <- length(unique(Mass_Samples_accCodes$SequenceCluster))
mass_VT <- rep(0, mass_clusters)

for (i in 1:mass_clusters){
  mass_VT[i] <- ceiling(median(as.integer(Mass_Samples_accCodes[Mass_Samples_accCodes$SequenceCluster == i-1,"Vaccine Type"]=="VT")))
}
# 1 means VT
# 0 means NVT
# if cluster is 50/50 it will count as VT
```

```{r}
#save VTs
#saveRDS(mass_VT, file = "mass_VT.rds")
```

```{r}
#calculate the frequency of the gene clusters and year
mass_cluster_freq_1 <- rep(0,mass_clusters)
mass_cluster_freq_2 <- rep(0,mass_clusters)
mass_cluster_freq_3 <- rep(0,mass_clusters)
for (i in 1:mass_clusters){
  mass_cluster_freq_1[i] <- length(which(Mass_Samples_accCodes[which(Mass_Samples_accCodes$SequenceCluster==i-1),]$"Year of Isolation"==2001))
  mass_cluster_freq_2[i] <- length(which(Mass_Samples_accCodes[which(Mass_Samples_accCodes$SequenceCluster==i-1),]$"Year of Isolation"==2004))
  mass_cluster_freq_3[i] <- length(which(Mass_Samples_accCodes[which(Mass_Samples_accCodes$SequenceCluster==i-1),]$"Year of Isolation"==2007))
}
```

```{r}
#save sequence cluster frequencies
#saveRDS(mass_cluster_freq_1, file = "mass_cluster_freq_1.rds")
#saveRDS(mass_cluster_freq_2, file = "mass_cluster_freq_2.rds")
#saveRDS(mass_cluster_freq_3, file = "mass_cluster_freq_3.rds")
```

```{r}
# For strong and weak selection in the model, I need to calculate the beta statistics.

# calculate gene frequencies first, separate for three time points
mass_gene_freq_1 <- apply(intermed_gene_presence_absence_2001[-1,-1], 1, sum_as_int)
mass_gene_freq_2 <- apply(intermed_gene_presence_absence_2004[-1,-1], 1, sum_as_int)
mass_gene_freq_3 <- apply(intermed_gene_presence_absence_2007[-1,-1], 1, sum_as_int)

# first, calculate pre/peri and post vacc frequencies of genes:
pre_peri_vacc_gene_freq <- (mass_gene_freq_1 + mass_gene_freq_2) / (sum(mass_cluster_freq_1) + sum(mass_cluster_freq_2))
pre_vacc_gene_freq <- (mass_gene_freq_1) / sum(mass_cluster_freq_1)
post_vacc_gene_freq <- mass_gene_freq_3 / sum(mass_cluster_freq_3)
peri_post <- (mass_gene_freq_2 + mass_gene_freq_3) / (sum(mass_cluster_freq_2) + sum(mass_cluster_freq_3))

# calculate delta statistic (refer to Corander et al. for more info)
delta_data <- (post_vacc_gene_freq - pre_peri_vacc_gene_freq) ^ 2 / (1 - pre_peri_vacc_gene_freq * (1 - pre_peri_vacc_gene_freq))
delta_ranking <- rank(delta_data)

delta_data2 <- (post_vacc_gene_freq - pre_vacc_gene_freq) ^ 2 / (1 - pre_vacc_gene_freq * (1 - pre_vacc_gene_freq))
delta_ranking2 <- rank(delta_data2)

delta_data3 <- (peri_post - pre_vacc_gene_freq) ^ 2 / (1 - pre_vacc_gene_freq * (1 - pre_vacc_gene_freq))
delta_ranking3 <- rank(delta_data3)
delta_ranking <- delta_ranking3
```

```{r}
#save delta ranking
#saveRDS(delta_ranking, file = "delta_ranking.rds")
```

```{r}
### create initial population that is based on the 2001 data set but not an exact sampling from it
# but a Poisson process
expand_factor <- 15000 / sum(mass_cluster_freq_1)
exp_noise <- 10
model_start_pop <- (sapply((mass_cluster_freq_1 + rexp(n = length(mass_cluster_freq_1), rate = exp_noise)) * expand_factor, rpois, n=1))

# visual check for the "sampling"
plot(mass_cluster_freq_1/sum(mass_cluster_freq_1))
points(model_start_pop/sum(model_start_pop), col = "red")
# looks similar enough, I would say
```

```{r}
#save model start population
#saveRDS(model_start_pop, file = "model_start_pop.rds")
```


### PopPUNK

# Read in PopPUNK clusters:
```{r}
PopPUNK_clusters <- read.csv("~/Documents/PhD_Project/Code/1st_project/odin-dust-examples/Data/refined_modelfitk3_clusters.csv")
no_PopPUNK_clusters <- length(unique(PopPUNK_clusters$Cluster)) # number of clusters in dataset
```

```{r}
# Add Isolate Name to PopPUNK data frame
accNo_to_filename <- read.delim("~/Documents/PhD_Project/Code/1st_project/odin-dust-examples/Data/filereport_read_run_PRJEB2632_tsv.txt")
accNo_to_filename <- accNo_to_filename[,c(1,8)]
accNo_to_filename[,3] <- matrix(unlist(strsplit(accNo_to_filename[,2],"/")), ncol=6, byrow = TRUE)[,6]
accNo_to_filename[,3] <- matrix(unlist(strsplit(accNo_to_filename[,3],"[.]")), ncol=2, byrow = TRUE)[,1]
accNo_to_filename <- accNo_to_filename[,c(1,3)]
colnames(accNo_to_filename) <- c(colnames(accNo_to_filename)[1], "filenames")

filenameToAccNo_dict <- accNo_to_filename$run_accession
names(filenameToAccNo_dict) <- accNo_to_filename$filenames

AccNoToIsolate_dict <- Mass_Samples_accCodes$`Isolate Name`
names(AccNoToIsolate_dict) <- Mass_Samples_accCodes$`Accession Code`

PopPUNK_clusters$IsolateName <- AccNoToIsolate_dict[filenameToAccNo_dict[PopPUNK_clusters$Taxon]]
```

```{r}
# Add VT information
VT_dict <- Mass_Samples_accCodes$`Vaccine Type`
names(VT_dict) <- Mass_Samples_accCodes$`Isolate Name`

PopPUNK_clusters$VT <- VT_dict[PopPUNK_clusters$IsolateName]
```

```{r}
# Add sequencing year information
SeqYear_dict <- Mass_Samples_accCodes$`Year of Isolation`
names(SeqYear_dict) <- Mass_Samples_accCodes$`Isolate Name`

PopPUNK_clusters$SeqYear <- SeqYear_dict[PopPUNK_clusters$IsolateName]
```


```{r}
#save model start population
#saveRDS(PopPUNK_clusters, file = "PopPUNK_clusters.rds")
PopPUNK_clusters <- readRDS(file = "PopPUNK_clusters.rds")
```

```{r}
#Compute consensus genomes for sequence clusters
PP_Mass_seq_clusters_dict <- PopPUNK_clusters$Cluster
names(PP_Mass_seq_clusters_dict) <- PopPUNK_clusters$IsolateName

PP_intermed_gene_presence_absence_consensus <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(unique(PopPUNK_clusters$Cluster))+1))
PP_intermed_gene_presence_absence_consensus[1,-1] <- paste("SeqCl_",unique(PopPUNK_clusters$Cluster),sep = "")

PP_intermed_gene_presence_absence_consensus_2001 <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(unique(PopPUNK_clusters$Cluster))+1))
PP_intermed_gene_presence_absence_consensus_2001[1,-1] <- paste("SeqCl_",unique(PopPUNK_clusters$Cluster),sep = "")

PP_intermed_gene_presence_absence_consensus_2004 <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(unique(PopPUNK_clusters$Cluster))+1))
PP_intermed_gene_presence_absence_consensus_2004[1,-1] <- paste("SeqCl_",unique(PopPUNK_clusters$Cluster),sep = "")

PP_intermed_gene_presence_absence_consensus_2007 <- data.frame(matrix(0, nrow = sum(cog_filter)+1, ncol = length(unique(PopPUNK_clusters$Cluster))+1))
PP_intermed_gene_presence_absence_consensus_2007[1,-1] <- paste("SeqCl_",unique(PopPUNK_clusters$Cluster),sep = "")


cons_genomes <- function(x){
  as.double(ceiling(median(as.integer(x))))
}

for (i in unique(PopPUNK_clusters$Cluster)) {
  PP_intermed_gene_presence_absence_consensus[-1,i+1] <- apply(as.matrix(intermed_gene_presence_absence[-1,c(FALSE,PP_Mass_seq_clusters_dict[unlist(intermed_gene_presence_absence[1,-1])]==i)]), 1, cons_genomes)
  PP_intermed_gene_presence_absence_consensus_2001[-1,i+1] <- apply(as.matrix(intermed_gene_presence_absence_2001[-1,c(FALSE,PP_Mass_seq_clusters_dict[unlist(intermed_gene_presence_absence_2001[1,-1])]==i)]), 1, cons_genomes)
    PP_intermed_gene_presence_absence_consensus_2004[-1,i+1] <- apply(as.matrix(intermed_gene_presence_absence_2004[-1,c(FALSE,PP_Mass_seq_clusters_dict[unlist(intermed_gene_presence_absence_2004[1,-1])]==i)]), 1, cons_genomes)
    PP_intermed_gene_presence_absence_consensus_2007[-1,i+1] <- apply(as.matrix(intermed_gene_presence_absence_2007[-1,c(FALSE,PP_Mass_seq_clusters_dict[unlist(intermed_gene_presence_absence_2007[1,-1])]==i)]), 1, cons_genomes)
}

PP_intermed_gene_presence_absence_consensus[-1,1] <- intermed_gene_presence_absence[-1,1]
PP_intermed_gene_presence_absence_consensus_2001[-1,1] <- intermed_gene_presence_absence[-1,1]
PP_intermed_gene_presence_absence_consensus_2004[-1,1] <- intermed_gene_presence_absence[-1,1]
PP_intermed_gene_presence_absence_consensus_2007[-1,1] <- intermed_gene_presence_absence[-1,1]
```


```{r}
#save intermed gene presence absence matrices
#saveRDS(PP_intermed_gene_presence_absence_consensus, file = "PP_intermed_gene_presence_absence_consensus.rds")
#saveRDS(PP_intermed_gene_presence_absence_consensus_2001, file = "PP_intermed_gene_presence_absence_consensus_2001.rds")
#saveRDS(PP_intermed_gene_presence_absence_consensus_2004, file = "PP_intermed_gene_presence_absence_consensus_2004.rds")
#saveRDS(PP_intermed_gene_presence_absence_consensus_2007, file = "PP_intermed_gene_presence_absence_consensus_2007.rds")
```

```{r}
# read in gene presence absence matrix
PP_intermed_gene_presence_absence_consensus <- readRDS(file = "PP_intermed_gene_presence_absence_consensus.rds")
PP_intermed_gene_presence_absence_consensus_2001 <- readRDS(file = "PP_intermed_gene_presence_absence_consensus_2001.rds")
PP_intermed_gene_presence_absence_consensus_2004 <- readRDS(file = "PP_intermed_gene_presence_absence_consensus_2004.rds")
PP_intermed_gene_presence_absence_consensus_2007 <- readRDS(file = "PP_intermed_gene_presence_absence_consensus_2007.rds")
```

```{r}
# calculate Vaccine Type consensus for clusters
PP_mass_VT <- rep(0, no_PopPUNK_clusters)

for (i in 1:no_PopPUNK_clusters){
  PP_mass_VT[i] <- ceiling(median(as.integer(PopPUNK_clusters[PopPUNK_clusters$Cluster == i,"VT"]=="VT")))
}
# 1 means VT
# 0 means NVT
# if cluster is 50/50 it will count as VT
```

```{r}
#save VTs
#saveRDS(PP_mass_VT, file = "PP_mass_VT.rds")
```

```{r}
#calculate the frequency of the gene clusters and year
PP_mass_cluster_freq_1 <- rep(0,no_PopPUNK_clusters)
PP_mass_cluster_freq_2 <- rep(0,no_PopPUNK_clusters)
PP_mass_cluster_freq_3 <- rep(0,no_PopPUNK_clusters)
for (i in 1:no_PopPUNK_clusters){
  PP_mass_cluster_freq_1[i] <- length(which(PopPUNK_clusters[which(PopPUNK_clusters$Cluster==i),]$"SeqYear"==2001))
  PP_mass_cluster_freq_2[i] <- length(which(PopPUNK_clusters[which(PopPUNK_clusters$Cluster==i),]$"SeqYear"==2004))
  PP_mass_cluster_freq_3[i] <- length(which(PopPUNK_clusters[which(PopPUNK_clusters$Cluster==i),]$"SeqYear"==2007))
}
```

```{r}
#save sequence cluster frequencies
#saveRDS(PP_mass_cluster_freq_1, file = "PP_mass_cluster_freq_1.rds")
#saveRDS(PP_mass_cluster_freq_2, file = "PP_mass_cluster_freq_2.rds")
#saveRDS(PP_mass_cluster_freq_3, file = "PP_mass_cluster_freq_3.rds")
```

```{r}
### create initial population that is based on the 2001 data set but not an exact sampling from it
# but a Poisson process
PP_expand_factor <- 15000 / sum(PP_mass_cluster_freq_1)
exp_noise <- 10
PP_model_start_pop <- (sapply((PP_mass_cluster_freq_1 + rexp(n = length(PP_mass_cluster_freq_1), rate = exp_noise)) * PP_expand_factor, rpois, n=1))

# visual check for the "sampling"
plot(PP_mass_cluster_freq_1/sum(PP_mass_cluster_freq_1))
points(PP_model_start_pop/sum(PP_model_start_pop), col = "red")
# looks similar enough, I would say
```

```{r}
#save model start population
#saveRDS(PP_model_start_pop, file = "PP_model_start_pop.rds")
```


### ggCaller

```{r}
# reading in the gene presence absence matrix produced by ggCaller
ggCaller_gene_presence_absence <- read.csv(paste(path_to_data, "Massachusetts_ggcaller/run2/ggCaller_output/gene_presence_absence.csv", sep = ""), header=FALSE)

# converting the gene presence absence matrix into a boolean df (0 = gene not present, 1 = gene present)
convert_to_bool <- function(x){
  if (x=="") 0 else 1
}

ggCaller_bool_gene_presence_absence <- ggCaller_gene_presence_absence[,c(-2,-3)]
ggCaller_bool_gene_presence_absence[-1,-1] <- apply(ggCaller_bool_gene_presence_absence[-1,-1],c(1,2), convert_to_bool)
```

