---
title: "Wright Fisher Model for 1 Allele and a haploid population"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{sir_models}
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: inline
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
lang_output <- function(x, lang) {
  cat(c(sprintf("```%s", lang), x, "```"), sep = "\n")
}
cc_output <- function(x) lang_output(x, "cc")
r_output <- function(x) lang_output(x, "r")
plain_output <- function(x) lang_output(x, "plain")
```

# Stochastic Lotka Volterra model definition


$$\begin{align*}
\frac{dR}{dt} &= \alpha R -\beta RC \\
\frac{dC}{dt} &= \delta \beta RC - \gamma C \\
\end{align*}$$

$R$ is the number of prey, $C$ is the number of predators.$\alpha$ is the birth rate of the prey, $\beta$ is the kill rate, $\gamma$ is the death rate of the predators and $\delta$ describes how much a predator benefits from killing prey.

Discretising this model in time steps of width $dt$ gives the following update equations for each time step:

$$\begin{align*}
R_{t+1} &= R_t - n_{RC} + n_R \\
C_{t+1} &= C_t + \delta n_{RC}  - n_C
\end{align*}$$

where
$$\begin{align*}
n_{R} &\sim B(R, 1 - e^{-\alpha \cdot dt}) \\
n_{RC} &\sim B(R, 1 - e^{-\beta C \cdot dt}) \\
n_{C} &\sim B(C, 1 - e^{-\gamma \cdot dt})
\end{align*}$$

## Implementing the SIR model using [`odin.dust`](https://mrc-ide.github.io/odin.dust/)

The above equations can straightforwardly be written out using the odin DSL:

```{r odin_sir, echo = FALSE, results = "asis"}
r_output(readLines(file.path("WrightFisher_nGenotypes_haploid_PopsizeVariablePois.R")))
```

This is converted to a C++ dust model, and compiled into a library in a single step, using [`odin.dust`](https://mrc-ide.github.io/odin.dust/). Save the above code as a file named `sir.R`. File names must not contain special characters.

```{r}
# install.packages("drat") # -- if you don't have drat installed
# drat:::add("ncov-ic")
# install.packages("odin.dust")
library(odin.dust)
```

```{r}
WF_nG_h_vP <- odin.dust::odin_dust("WrightFisher_nGenotypes_haploid_PopsizeVariablePois.R")
```

```{r}
# add low and high levels of selection to the model.

# first, calculate pre/peri and post vacc frequencies of genes:
pre_vacc_pop <- c(100,300,300,100)
genotype_matr <- cbind(c(1,1,0,0,0),c(0,0,1,0,0),c(1,0,0,1,1),c(0,0,0,1,1))
pre_vacc_freq <- c(4/8, 1/8, 3/8, 4/8, 4/8) # calculated that manually from pre_vacc_pop and genotype_matr
post_vacc_freq <- c(3/8, 1/8, 4/8, 4/8, 4/8) # just an example
is_vaccType <- c(0,0,1,0)

# calculate delta statistic (refer to Corander et al. for more info)
delta_ex <- (post_vacc_freq - pre_vacc_freq) ^ 2 / (1 - pre_vacc_freq * (1 - pre_vacc_freq))
delta_rank <- rank(delta_ex)
#<- sort(delta_data, index.return = TRUE)
# access the indexes of the corresponding genes through sorted_delta$ix, access the values through sorted_delta$x

```

### Define model parameters
```{r}
#params_n_vP <- list(dt = 1, species_no = 4,  gene_no = 5, Pop_ini = rep(100,4), Pop_eq = c(400, 20, 20, 20), capacity = 400, Genotypes = cbind(c(1,0,0,0,0),c(0,0,1,0,0),c(1,0,0,1,1),c(0,0,0,0,1)), GeneFitness = c(0.02, 0, 0, 0.01, 0.03)) 
# the following parameters work reliably to produce the expected behaviour:
#params_n_vP <- list(dt = 1, species_no = 4,  gene_no = 5, Pop_ini = c(150,350,350,150), Pop_eq = c(100,400,400,100), capacity = 1000, Genotypes = cbind(c(1,0,0,0,0),c(0,0,1,0,0),c(1,0,0,1,1),c(0,0,0,1,1)), sigma = 0.02, m = 0.002) 
#params_n_vP <- list(dt = 1, species_no = 4,  gene_no = 5, Pop_ini = c(150,350,350,150), Pop_eq = c(100,400,400,100), capacity = 1000, Genotypes = cbind(c(1,0,0,0,0),c(0,0,1,0,0),c(1,0,0,1,1),c(0,0,0,1,1)), sigma = 0.02, m = 0.00, vaccTypes = c(0,0,1,0), v = 0.002, vacc_time = 100) 
# the following parameters create the setting where two strains survive and two die!
#params_n_vP <- list(dt = 1/36, species_no = 4,  gene_no = 5, Pop_ini = c(150,350,350,150), Pop_eq = c(100,400,400,100), capacity = 1000, Genotypes = cbind(c(1,0,0,0,0),c(0,0,1,0,0),c(1,0,0,1,1),c(0,0,0,1,1)), sigma = 0.02, m = 0.00, vaccTypes = c(0,0,1,0), v = 0, vacc_time = 100) 
params_n_vP <- list(dt = 1/36, species_no = 4,  gene_no = 5, Pop_ini = pre_vacc_pop, Pop_eq = pre_vacc_pop, capacity = sum(pre_vacc_pop), Genotypes = genotype_matr, sigma_f = 0.14, sigma_w = 0.002, prop_f = 0.25, delta = delta_rank, m = 0.03, vaccTypes = is_vaccType, v = 0.07, vacc_time = 5) 
# the following parameters show the expected behaviour in about 50% of the times. Probably stochastic effects because the population sizes of gt 1, 3 and 4 are so small in the beginning?
# params_n_vP <- list(dt = 1, species_no = 4,  gene_no = 5, Pop_ini = c(150,550,250,150), Pop_eq = c(100,400,400,100), capacity = 1000, Genotypes = cbind(c(1,0,0,0,0),c(0,0,1,0,0),c(1,0,0,1,1),c(0,0,0,1,1)), sigma = 0.02)
### maybe I can investigate whether binomial draws are more robust? (later)

#params_n_vP <- list(dt = 1, species_no = 4,  gene_no = 5, Pop_ini = rep(100,4), Pop_eq = c(20, 20, 400, 20), capacity = 300, Genotypes = cbind(c(1,0,0,0,0),c(0,1,0,0,0),c(1,0,0,1,1),c(0,0,0,0,1)), GeneFitness = c(0.1, 0.1, 0.2, 0.1, 0.1))
```


### Running the model:
```{r}
WFmodel_nG_h_vP <- WF_nG_h_vP$new(pars = params_n_vP,
                         time = 1,
                         n_particles = 10L,
                         n_threads = 4L,
                         seed = 1L)
```

# initial state: 
```{r}
WFmodel_nG_h_vP$state()
```


# running the model for 10, 20, 30 time steps:
```{r}
WFmodel_nG_h_vP$run(1)
WFmodel_nG_h_vP$run(2)
WFmodel_nG_h_vP$run(3)
```



```{r fig.height=5, fig.width=7}
dt <- 1
n_particles <- 10L
#WFmodel_nG_h_vP$update_state(pars = params_n_vP, time = 0)
WFmodel_nG_h_vP <- WF_nG_h_vP$new(pars = params_n_vP,
                         time = 1,
                         n_particles = 10L,
                         n_threads = 4L,
                         seed = 1L)


n_times <- 1000
x <- array(NA, dim = c(WFmodel_nG_h_vP$info()$len, n_particles, n_times))

for (t in seq_len(n_times)) {
  x[ , , t] <- WFmodel_nG_h_vP$run(t)
}
time <- x[1, 1, ]
x <- x[-1, , ]

par(mar = c(4.1, 5.1, 0.5, 0.5), las = 1)
if (length(x[,1,1]) <= 8){
  cols <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7") #8 colorblind friendly colors
} else {
  cols <- rainbow(length(x[,1,1]))
}

library(dplyr)
library(zoo)

matplot(time, rollmean(x[1,1 , ], k=30, fill=NA), type = "l",
         xlab = "Time", ylab = "Number of individuals",
         col = cols[1], lty = 1, ylim = range(x))
for (species in 2:length(x[,1,1])) {
  matlines(time, rollmean(x[species,1 , ], k=30, fill=NA), col = cols[species], lty = 1)
}

#matplot(time, (x[1,1 , ]), type = "l",
#         xlab = "Time", ylab = "Number of individuals",
#         col = cols[1], lty = 1, ylim = range(x))
#for (species in 2:length(x[,1,1])) {
#  matlines(time, (x[species,1 , ]), col = cols[species], lty = 1)
#}



legend("bottomright", lwd = 1, col = cols[1:length(x[,1,1])], legend = 1:length(x[,1,1]), bty = "n")
```

```{r fig.height=5, fig.width=7}
matplot(time, t(x[1, , ]), type = "l",
         xlab = "Time", ylab = "Number of individuals",
         col = cols[1], lty = 1, ylim = range(x))
for (species in 2:length(x[,1,1])) {
  matlines(time, t(x[species,, ]), col = cols[species], lty = 1)
}
#matlines(time, t(x[species,, ]), col = cols[species], lty = 1)

legend("right", lwd = 1, col = cols,  legend = 1:length(x[,1,1]), bty = "n")
```

```{r fig.height=5, fig.width=7}
dt <- 1
n_particles <- 10L
#WFmodel_nG_h_vP$update_state(pars = params_n_vP, time = 0)
WFmodel_nG_h_vP <- WF_nG_h_vP$new(pars = params_n_vP,
                         time = 1,
                         n_particles = 10L,
                         n_threads = 4L,
                         seed = 1L)


n_times <- 10000
x <- array(NA, dim = c(WFmodel_nG_h_vP$info()$len, n_particles, n_times))

for (t in seq_len(n_times)) {
  x[ , , t] <- WFmodel_nG_h_vP$run(t)
}
time <- x[1, 1, ]
x <- x[-1, , ]

par(mar = c(4.1, 5.1, 0.5, 0.5), las = 1)
if (length(x[,1,1]) <= 8){
  cols <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7") #8 colorblind friendly colors
} else {
  cols <- rainbow(length(x[,1,1]))
}

# probably should find a more dynamic way for models with >8 species?
matplot(time, (x[1,2 , ]), type = "l",
         xlab = "Time", ylab = "Number of individuals",
         col = cols[1], lty = 1, ylim = range(x))
for (species in 2:length(x[,2,1])) {
  matlines(time, (x[species,2 , ]), col = cols[species], lty = 1)
}



legend("right", lwd = 1, col = cols[1:length(x[,2,1])], legend = 1:length(x[,2,1]), bty = "n")
```

```{r}
# create plot colored by whether clusters are vaccine types or not

par(mar = c(4.1, 5.1, 0.5, 0.5), las = 1)
cols <- rep("#000000", length(x[,1,1]))
for (i in 1:length(cols)){
  cols[i] <- if(is_vaccType[i]==1){"#E69F00"} else {"#000000"}
}


# probably should find a more dynamic way for models with >8 species?
matplot(time, rollmean(x[1,1 , ], k=30, fill=NA), type = "l",
         xlab = "Time", ylab = "Number of individuals",
         col = cols[1], lty = 1, ylim = range(x))
for (species in 2:length(x[,1,1])) {
  matlines(time, rollmean(x[species,1 , ], k=30, fill=NA), col = cols[species], lty = 1)
}

# or 
# matplot(time,(x[1,1 , ]), type = "l",
#         xlab = "Time", ylab = "Number of individuals",
#         col = cols[1], lty = 1, ylim = range(x))
#for (species in 2:length(x[,1,1])) {
#  matlines(time, (x[species,1 , ]), col = cols[species], lty = 1)
#}
#if you want to see all single values, not the monthly rolling mean


legend("topright", lwd = 1, col = c("#000000","#E69F00"), legend = c("Vaccine Types", "Non-Vaccine Types"), bty = "n")

# vaccine types seem to be less successful, yeah!
```

```{r}
# make a grouped bar plot similar to Croucher et al.
library(ggplot2)

no_clusters = 4
cluster_names <- rep(1:no_clusters, 3)
cluster_names <- sort(cluster_names)
#cluster_names <- as.character(cluster_names)
simTimes <- rep(c("pt 1", "pt 2", "pt 3"), no_clusters)
sim_cluster_freq_all_times <- rep(0, no_clusters *3)
for (i in 1:no_clusters) {
  sim_cluster_freq_all_times[3*(i-1) + 1 ] <- x[i,1,1]/sum(x[,1,1])
  sim_cluster_freq_all_times[3 * (i-1) + 2] <- x[i,1,n_times/2]/sum(x[,1,n_times/2])
  sim_cluster_freq_all_times[3 * (i-1) + 3] <- x[i,1,n_times]/sum(x[,1,n_times])
}
sim_cluster_freq_df <- data.frame(cluster_names, simTimes, sim_cluster_freq_all_times)


library(gridExtra)


ggplot(sim_cluster_freq_df, aes(fill=simTimes, y=sim_cluster_freq_all_times, x=cluster_names)) +   geom_bar(position="dodge", stat="identity") +
scale_fill_manual("legend", values = c("pt 1" = "#E69F00", "pt 2" = "#56B4E9", "pt 3" = "#009E73")) +
scale_x_continuous(breaks = 1:no_clusters)
```

### now, the fitting can begin! :-)
```{r}
#install.packages("drat") # -- if you don't have drat installed
drat:::add("ncov-ic")
#install.packages("mcstate")
library(mcstate)
```

```{r}
# create example data set
test_data <- data.frame("year" = c(2001, 2004, 2007), "1" = c(100, 300, 400), "2" = c(300, 200, 300), "3" = c(300, 100, 50), "4" = c(100, 200, 300))
names(test_data) <- c("year", "1", "2", "3", "4")
```

```{r}
# process data with particle filter:
dt <- 1/36 # we assume that the generation time of Strep. pneumo is 1 month
# we have data from 2001, 2004, 2007, so we want 3 (years) * 12 (months) = 36 updates in-between
test_data_filt <- mcstate::particle_filter_data(data = test_data,
                                          time = "year",
                                          rate = 1 / dt,
                                          initial_time = 2001)
```

```{r}
# Defining the comparison function
# state is the model state at some point in time
# observed is the real data 
cluster_compare <- function(state, observed, pars = NULL) {
  exp_noise <- 1e6
  print(dim(state))
  
  #state_x[ , ] <- state
  #state <- state[-1, ]
  clusters_modelled <- state[2, , drop = TRUE]
  print(clusters_modelled)
  clusters_observed <- observed$"1"
  lambda <- clusters_modelled +
    rexp(n = length(clusters_modelled), rate = exp_noise)
  dpois(x = clusters_observed, lambda = lambda, log = TRUE)
}

```

```{r}
# heavily based on this: https://mrc-ide.github.io/mcstate/articles/sir_models.html
# trying to generalise the comparison function so that we can later fit to all strain data
# log-likelihood of Poisson count
ll_pois <- function(obs, model) {
  exp_noise <- 1e6
  if (is.na(obs)) {
    # Creates vector of zeros in ll with same length, if no data
    ll_obs <- numeric(length(model))
  } else {
    lambda <- model +
      rexp(n = length(model), rate = exp_noise)
    ll_obs <- dpois(x = obs, lambda = lambda, log = TRUE)
  }
  ll_obs
}

combined_compare <- function(state, observed, pars = NULL) {
  ll_one <- ll_pois(observed$"1", state[2, , drop = TRUE])
  ll_two <- ll_pois(observed$"2", state[3, , drop = TRUE])
  ll_three <- ll_pois(observed$"3", state[4, , drop = TRUE])
  ll_four <- ll_pois(observed$"4", state[5, , drop = TRUE])
  ll_one + ll_two + ll_three + ll_four
}
```

```{r}
### Inferring parameters

n_particles <- 100
### I have an issue with the dimensions!
# these three blocks of code only run if n_particles = 62.
# so probably it confuses the number of strains and the number of particles.
# that is a bit weird, I think.
# maybe in my data, it think that the three times points are the three compartments?
# but that does not really make sense - it recognizes 2001, 2004 and 2007 as times.
# BUT the info about the different strains got lost - it is not in the dataset.
# SO that is because for the SIR they only had the data for one compartment, the I, here I have infos about all compartments.
filter <- mcstate::particle_filter$new(data = test_data_filt,
                                       model = WF_nG_h_vP,
                                       n_particles = n_particles,
                                       compare = combined_compare,
                                       seed = 1L)
```

```{r}
filter$run(save_history = TRUE, pars = params_n_vP)
# currently only filtering strain 1
```

```{r}
# plot particle filter

plot_particle_filter <- function(history, true_history, times, obs_end = NULL) {
  if (is.null(obs_end)) {
    obs_end <- max(times)
  }

  par(mar = c(4.1, 5.1, 0.5, 0.5), las = 1)
  cols <- c("1" = "#8c8cd9", "2" = "#cc0044", "3" = "#999966", "4" = "orange")
  
  matplot(times, t(history[2, , -1]), type = "l",
          xlab = "Time", ylab = "Number of individuals",
          col = cols[["1"]], lty = 1, ylim = range(history[2:5, , -1]))
  matlines(times, t(history[3, , -1]), col = cols[["2"]], lty = 1)
  matlines(times, t(history[4, , -1]), col = cols[["3"]], lty = 1)
  matlines(times, t(history[5, , -1]), col = cols[["4"]], lty = 1)
  print(nrow(t(true_history[1:3, , -1])))
  matpoints(times[1:obs_end], (true_history[1:3,2 , -1]), pch = 19,
            col = cols[["1"]])
  matpoints(times[1:obs_end], (true_history[1:3,3 , -1]), pch = 19,
            col = cols[["2"]])
  matpoints(times[1:obs_end], (true_history[1:3,4 , -1]), pch = 19,
            col = cols[["3"]])
  matpoints(times[1:obs_end], (true_history[1:3,5 , -1]), pch = 19,
            col = cols[["4"]])
  legend("left", lwd = 1, col = cols, legend = names(cols), bty = "n")
}

true_history <- test_data
plot_particle_filter(filter$history(), true_history, 1:3)
```

```{r}
# okay, it's running now.
# next step is to filter all strain data not just against the data of strain 1.

```

```{r}
# Using MCMC to infer parameters
pmcmc_sigma_f <- mcstate::pmcmc_parameter("sigma_f", 0.15, min = 0)
pmcmc_sigma_w <- mcstate::pmcmc_parameter("sigma_w", 0.05, min = 0)
pmcmc_prop_f <- mcstate::pmcmc_parameter("prop_f", 0.25, min = 0, max = 1)
pmcmc_m <- mcstate::pmcmc_parameter("m", 0.03, min = 0)
pmcmc_v <- mcstate::pmcmc_parameter("v", 0.05, min = 0)
species_no <- 4
gene_no <- 5
Pop_ini <- pre_vacc_pop
Pop_eq <- pre_vacc_pop
Genotypes <-genotype_matr
capacity <- sum(pre_vacc_pop)
delta <- delta_rank
vaccTypes <- is_vaccType
vacc_time <- 5

#make_transform <- function(m) {
#  function(theta) {
#    c(list(m = m), as.list(theta))
#  }
#}

  ## The transform management is hardest; need to provide a *list*
      ## of functions, one per population.
      #make_transform <- function(pop) {
       # base <- base[, pop, drop = TRUE]
       # base_transform <- private$transform[[pop]]
       # function(p) {
       #   base[i_keep] <- p
       #   base_transform(base)
       # }
     # }
     # base <- matrix(NA_real_, length(self$names()), length(populations),
      #               dimnames = list(self$names(), populations))
    #  base[rownames(fixed), ] <- fixed
complex_params <- c(Pop_ini, Pop_eq, Genotypes, capacity, delta, vaccTypes, species_no, gene_no, vacc_time, 0.15, 0.05, 0.25, 0.03, 0.05)
#set_names <- function(x, nms) {
#  names(x) <- nms
#  x
#}
#transform <- set_names(lapply(complex_params, make_transform), complex_params)
make_transform <- function(p) {
    list(Pop_ini = p[1:4],
         Pop_eq = p[5:8],
         Genotypes = matrix(p[9:28], nrow = gene_no),
         capacity = p[29],
         delta = p[30:34],
         vaccTypes = p[35:38],
         species_no = p[39],
         gene_no = p[40],
         vacc_time = p[41],
         sigma_f = p[42],
         sigma_w = p[43],
         prop_f = p[44],
         m = p[45],
         v = p[46])
  }

transform <- function(x) {
  make_transform(complex_params)}
proposal_matrix <- diag(0.1, 5) # the proposal matrix defines the covariance-variance matrix for a mult normal dist
# here, all parameters are proposed independently. 
# think about this, this might not actually be true
#mcmc_pars <- mcstate::pmcmc_parameters$new(list(pmcmc_sigma_f, pmcmc_sigma_w, pmcmc_prop_f, pmcmc_m, pmcmc_v), proposal_matrix, transform)
mcmc_pars <- mcstate::pmcmc_parameters$new(list(mcstate::pmcmc_parameter("sigma_f", 0.15, min = 0), mcstate::pmcmc_parameter("sigma_w", 0.05, min = 0), mcstate::pmcmc_parameter("prop_f", 0.25, min = 0, max = 1), mcstate::pmcmc_parameter("m", 0.03, min = 0), mcstate::pmcmc_parameter("v", 0.05, min = 0)), proposal_matrix, transform)
#= make_transform(c(Pop_ini, Pop_eq, Genotypes, capacity, delta, vaccTypes, species_no, gene_no, vacc_time)))
mcmc_pars$names()
mcmc_pars$model(mcmc_pars$initial())
mcmc_pars$initial()
# read this: https://mrc-ide.github.io/mcstate/reference/pmcmc_parameters.html
# it explains how to not fit all parameters but just the ones I want
# non-scalar parameters have to be transformed for this.
```

```{r}
n_steps <- 500
n_burnin <- 200
```

```{r}
control <- mcstate::pmcmc_control(
    n_steps,
    save_state = TRUE,
    save_trajectories = TRUE,
    progress = TRUE)
pmcmc_run <- mcstate::pmcmc(mcmc_pars, filter, control = control)
# Running chain 1 / 1
# Error in filter_run(self, private, pars, save_history, save_restart, min_log_likelihood) : 
# could not find function "filter_run"
### I do not understand this error message.
### might be that I still have issues with my parameters :-/

plot_particle_filter(pmcmc_run$trajectories$state, true_history, 1:3)
```
```{r}
processed_chains <- mcstate::pmcmc_thin(pmcmc_run, burnin = n_burnin, thin = 2)
parameter_mean_hpd <- apply(processed_chains$pars, 2, mean)
parameter_mean_hpd
```

