---
title: "Navajo Data Processing"
output: html_notebook
editor_options: 
  chunk_output_type: console
---

```{r}
path_to_data <- "~/Documents/PhD_Project/Data/"
```

# read in meta data (esp. serotype, sequencing year)
```{r}
library(readxl)
Navajo_meta_data <- read_excel(paste(path_to_data, "StrepPneumo_Navajo/metadata.xlsx", sep = ""))
Navajo_time_points <- sort(unique(Navajo_meta_data$Year))

Navajo_SeqYear_dict <- Navajo_meta_data$Year
names(Navajo_SeqYear_dict) <- Navajo_meta_data$Accession

Navajo_meta_IDtoInd <- 1:nrow(Navajo_meta_data)
names(Navajo_meta_IDtoInd) <- Navajo_meta_data$Accession
```


```{r}
Navajo_time_seq <- Navajo_time_points[1]:Navajo_time_points[length(Navajo_time_points)]
seq_no_per_year <- rep(0, length(Navajo_time_seq))
names(seq_no_per_year) <- Navajo_time_seq
for (i in 1:length(Navajo_time_seq)) {
  seq_no_per_year[i] <- length(which(Navajo_meta_data$Year==Navajo_time_seq[i]))
}
barplot(seq_no_per_year, xlab = "Sequencing Years", ylab = "Sequence Counts", col = rainbow(15))
# PCV7 was introduced in 2000
# PCV13 was introduced in 2010
# according to Azarian et al. (2018) https://doi.org/10.1371/journal.ppat.1006966
# 
Navajo_time_points_selected <- Navajo_time_seq
Navajo_time_points_preVac1 <- Navajo_time_seq[1:3]
Navajo_time_points_preVac2 <- Navajo_time_seq[1:13]
```


# read in PopPUNK results
```{r}
PP_all <- read.csv(paste(path_to_data, "StrepPneumo_Navajo/poppunk_clusters/poppunk_clusters_clusters.csv",sep = ""))
PP_taxon_ind_dict <- 1:nrow(PP_all)
names(PP_taxon_ind_dict) <- PP_all$Taxon

PP_is_Navajo_dict <- rep(FALSE,nrow(PP_all))
names(PP_is_Navajo_dict) <- PP_all$Taxon

for (i in 1:length(Navajo_meta_data$Accession)) {
  PP_is_Navajo_dict[Navajo_meta_data$Accession[i]] <- TRUE
}
Navajo_PP <- PP_all[PP_is_Navajo_dict,]
no_Navajo_PP <- length(unique(Navajo_PP$Cluster)) 
# 74 clusters

Navajo_seq_clusters_dict <- Navajo_PP$Cluster
names(Navajo_seq_clusters_dict) <- Navajo_PP$Taxon
```

```{r}
saveRDS(Navajo_meta_data, "Navajo_meta_data.rds")
```

```{r}
# Add Serotype Information
Navajo_Serotype_dict <- Navajo_meta_data$Sero
names(Navajo_Serotype_dict) <- Navajo_meta_data$Accession

Navajo_PP$Serotype <- Navajo_Serotype_dict[Navajo_PP$Taxon]

saveRDS(Navajo_PP,"Navajo_PP.rds")
```


# read in ggCaller results
```{r}
Navajo_ggCaller <- read.csv(paste(path_to_data, "StrepPneumo_Navajo/gene_presence_absence.csv", sep = ""), header=FALSE)
```

```{r}
# converting the gene presence absence matrix into a boolean df (0 = gene not present, 1 = gene present)
convert_to_bool <- function(x){
  if (x=="") 0 else 1
}

Navajo_ggCaller_bool <- Navajo_ggCaller[,c(-2,-3)]
Navajo_ggCaller_bool[-1,-1] <- apply(Navajo_ggCaller_bool[-1,-1],c(1,2), convert_to_bool)
#Navajo_ggCaller_bool[1,-1] <- unlist(strsplit(Navajo_ggCaller_bool[1,-1],".contigs_velvet"))

Navajo_ggCaller_bool[1,-1] <- sapply(Navajo_ggCaller_bool[1,-1], strsplit, split = ".contigs_velvet") # remove contigs appendix
colnames(Navajo_ggCaller_bool) <- Navajo_ggCaller_bool[1,]

PP_taxon_in_ggC <- rep(FALSE, length(Navajo_PP$Taxon))
names(PP_taxon_in_ggC) <- Navajo_PP$Taxon

#> length(PP_taxon_in_ggC)
#[1] 937

for (i in 2:ncol(Navajo_ggCaller_bool)) {
  PP_taxon_in_ggC[colnames(Navajo_ggCaller_bool)[i]] <- TRUE
}

#> length(PP_taxon_in_ggC)
#[1] 951

# why??

# hm, maybe filter by accession numbers that I actually had?
# running PP in that --include_database mode really seems to risk that I include stuff that wasn't actually in Navajo...

ggC_has_meta <- rep(FALSE, (ncol(Navajo_ggCaller_bool)-1))
names(ggC_has_meta) <- colnames(Navajo_ggCaller_bool)[-1]
for (i in 1:nrow(Navajo_meta_data)) {
  ggC_has_meta[Navajo_meta_data$Accession[i]] <- TRUE
}

# there are 14 sequences in ggCaller for which I do not have any meta data. why?? 
# Well, okay, I only have metadate for 937 sequences. So I am only going to consider those.

meta_has_ggC <- rep(FALSE, nrow(Navajo_meta_data))
names(meta_has_ggC) <- Navajo_meta_data$Accession
for (i in 2:ncol(Navajo_ggCaller_bool)) {
  meta_has_ggC[colnames(Navajo_ggCaller_bool)[i]] <- TRUE
}

which(!meta_has_ggC)

which(!ggC_has_meta)
# filter ggCaller results by the ones 
Navajo_ggCallerPP_bool <- data.frame(matrix(0, nrow = nrow(Navajo_ggCaller_bool), ncol = length(Navajo_PP$Taxon)+1))
Navajo_ggCallerPP_bool[-1,-1] <- Navajo_ggCaller_bool[-1,(which(ggC_has_meta))]
colnames(Navajo_ggCallerPP_bool) <- c("Gene",Navajo_PP$Taxon)
Navajo_ggCallerPP_bool[1,] <- c("Gene",Navajo_PP$Taxon)
Navajo_ggCallerPP_bool[,1] <- Navajo_ggCaller_bool[,1]
```

```{r}
saveRDS(Navajo_ggCallerPP_bool, "Navajo_ggCaller_bool.rds")
```

# split ggCaller gene presence absence matrix into different time points
```{r}
Nepal_ggCaller_byYear <- vector(mode = "list", length = length(Nepal_time_points_selected))

for (i in 1:length(Nepal_time_points_selected)) {
  Nepal_ggCaller_year <- data.frame(matrix(0, nrow = nrow(Nepal_ggCallerPP_bool), ncol = length(which(Nepal_meta_dataPP$Year==Nepal_time_points_selected[i]))+1)) # initialize data frame
  Nepal_ggCaller_year[1,-1] <- Nepal_ggCallerPP_bool[1,c(FALSE,Nepal_SeqYear_dict[unlist(Nepal_ggCallerPP_bool[1,-1])]==Nepal_time_points_selected[i])] # fill in first row (seq names)
  Nepal_ggCaller_year[-1,1] <- Nepal_ggCallerPP_bool[-1,1] # fill in first column (gene cluster names)
  Nepal_ggCaller_year[-1,-1] <- Nepal_ggCallerPP_bool[-1,c(FALSE,Nepal_SeqYear_dict[unlist(Nepal_ggCallerPP_bool[1,-1])]==Nepal_time_points_selected[i])] # fill in presence absence information
  Nepal_ggCaller_byYear[[i]] <- Nepal_ggCaller_year
}
```

# find intermediate frequency genes
# I will combine all pre-vaccine data points for this
```{r}
sum_as_int <- function(x){
  sum(as.integer(x))
}

Nepal_gene_freq_preVac <- rep(0, nrow(Nepal_ggCaller_byYear[[1]])-1)
preVac_datapoints <- 0
for (i in 1:length(Nepal_time_points_preVac)) {
  Nepal_gene_freq_preVac <- Nepal_gene_freq_preVac + apply(Nepal_ggCaller_byYear[[i]][-1,-1],1, sum_as_int)
  preVac_datapoints <- preVac_datapoints + length(Nepal_ggCaller_byYear[[i]][1,-1])
}
Nepal_gene_freq_preVac <- Nepal_gene_freq_preVac / preVac_datapoints

Nepal_gene_filter <- as.integer(Nepal_gene_freq_preVac<=0.95 & Nepal_gene_freq_preVac>=0.05)
# 1987 intermediate-frequency genes (of 5775)
```

# create intermediate frequency dataframes
```{r}
Nepal_ggCaller_byYear_intermed <- vector(mode = "list", length = length(Nepal_time_points_selected))

for (i in 1:length(Nepal_time_points_selected)) {
  Nepal_ggCaller_year <- data.frame(matrix(0, nrow = sum(Nepal_gene_filter)+1,ncol = length(which(Nepal_meta_dataPP$Year==Nepal_time_points_selected[i]))+1))
  if(ncol(Nepal_ggCaller_byYear[[i]])>1){
    Nepal_ggCaller_year <-  Nepal_ggCaller_byYear[[i]][c(1,which(Nepal_gene_filter==1)+1),]
  }
  else{
    Nepal_ggCaller_year <-  data.frame(matrix(Nepal_ggCaller_byYear[[6]][c(1,which(Nepal_gene_filter==1)+1),1], nrow = sum(Nepal_gene_filter)+1,ncol = 1))
  }
  Nepal_ggCaller_byYear_intermed[[i]] <- Nepal_ggCaller_year 
}

Nepal_ggCaller_intermed <- data.frame(matrix(0, nrow = sum(Nepal_gene_filter)+1,ncol = ncol(Nepal_ggCallerPP_bool)))
Nepal_ggCaller_intermed <-  Nepal_ggCallerPP_bool[c(1,which(Nepal_gene_filter==1)+1),]

# create pre_Vac intermed frequency data frame
Nepal_ggCaller_intermed_preVac <- data.frame(matrix(0, nrow = sum(Nepal_gene_filter)+1,ncol = (sum(unlist(lapply(Nepal_ggCaller_byYear_intermed,ncol))[1:length(Nepal_time_points_preVac)]) - length(Nepal_time_points_preVac) + 1)))
col_count <- 2
for (i in 1:length(Nepal_time_points_preVac)) {
  if(ncol(Nepal_ggCaller_byYear_intermed[[i]])>1){
    Nepal_ggCaller_intermed_preVac[,(col_count:(col_count + ncol(Nepal_ggCaller_byYear_intermed[[i]])-1))] <- Nepal_ggCaller_byYear_intermed[[i]][,-1]
    col_count <- col_count + ncol(Nepal_ggCaller_byYear_intermed[[i]])
  }
}
Nepal_ggCaller_intermed_preVac[,1] <-  Nepal_ggCaller_byYear_intermed[[1]][,1]
```

# create consensus gene presence absence matrices for the clusters
```{r}
cons_genomes <- function(x){
  as.double(ceiling(median(as.integer(x))))
}

# overall
Nepal_ggCaller_intermed_consensus <- data.frame(matrix(0, nrow = sum(Nepal_gene_filter)+1, ncol = no_Nepal_PP+1))
Nepal_ggCaller_intermed_consensus[1,-1] <- paste("SeqCl_",unique(Nepal_PP$Cluster),sep = "")

for (i in unique(Nepal_PP$Cluster)) {
  Nepal_ggCaller_intermed_consensus[-1,(which(unique(Nepal_PP$Cluster) == i))+1] <- apply(as.matrix(Nepal_ggCaller_intermed[-1,c(FALSE,Nepal_seq_clusters_dict[unlist(Nepal_ggCaller_intermed[1,-1])]==i)]), 1, cons_genomes)

}
Nepal_ggCaller_intermed_consensus[,1] <- Nepal_ggCaller_intermed[,1]

#by year
Nepal_ggCaller_byYear_intermed_consensus <- vector(mode = "list", length = length(Nepal_time_points_selected))

for (i in 1:length(Nepal_time_points_selected)) {
  Nepal_ggCaller_year <- data.frame(matrix(0, nrow = sum(Nepal_gene_filter)+1,ncol = no_Nepal_PP+1))
  Nepal_ggCaller_year[1,-1] <- paste("SeqCl_",unique(Nepal_PP$Cluster),sep = "")
  for (j in unique(Nepal_PP$Cluster)) {
    Nepal_ggCaller_year[-1,(which(unique(Nepal_PP$Cluster) == j))+1] <- apply(as.matrix(Nepal_ggCaller_byYear_intermed[[i]][-1,c(FALSE,Nepal_seq_clusters_dict[unlist(Nepal_ggCaller_byYear_intermed[[i]][1,-1])]==j)]), 1, cons_genomes)
  }
  Nepal_ggCaller_year[,1] <- Nepal_ggCaller_byYear_intermed[[i]][,1]
  Nepal_ggCaller_byYear_intermed_consensus[[i]] <- Nepal_ggCaller_year
}

saveRDS(Nepal_ggCaller_intermed_consensus,"Nepal_ggCaller_intermed_consensus.rds")
```

# Calculate VT/NVT for clusters
```{r}
# create dictionary for VTs of the PCV10
# 1, 4, 5, 6b, 7f, 9V, 14, 18c, 19f and 23f serotypes (source: NCBI)
PCV10_VTs <- rep(0,length(unique(Nepal_meta_dataPP$Phenotypic_serotype)))
names(PCV10_VTs) <- unique(Nepal_meta_dataPP$Phenotypic_serotype)
PCV10_VTs[c("1","4","5","6B", "7F", "9V", "14", "18C", "19F", "23F")] <- 1

Nepal_SeqToSero_dict <- Nepal_meta_dataPP$Phenotypic_serotype
names(Nepal_SeqToSero_dict) <- Nepal_meta_dataPP$Lane_id

# calculate Vaccine Type consensus for clusters
Nepal_VT <- rep(0, no_Nepal_PP)

for (i in unique(Nepal_PP$Cluster)){
  Nepal_VT[(which(unique(Nepal_PP$Cluster) == i))] <- ceiling(median(as.integer(PCV10_VTs[Nepal_SeqToSero_dict[Nepal_PP[Nepal_PP$Cluster == i,"Taxon"]]])))
}
# 1 means VT
# 0 means NVT
# if cluster is 50/50 it will count as VT
```

```{r}
#save VTs
saveRDS(Nepal_VT, file = "Nepal_VT.rds")
```

# calculate cluster frequencies for different years
```{r}
Nepal_cluster_freqs <- vector(mode = "list", length = length(Nepal_time_points_selected))

for(j in 1:length(Nepal_time_points_selected)){
  Nepal_cluster_freqs[[j]] <- rep(0, no_Nepal_PP)
  for (i in unique(Nepal_PP$Cluster)){
    Nepal_cluster_freqs[[j]][which(unique(Nepal_PP$Cluster) == i)] <- length(which(Nepal_meta_dataPP[Nepal_meta_IDtoIndPP[Nepal_PP[Nepal_PP$Cluster == i,"Taxon"]],]$Year==Nepal_time_points_selected[j]))
  }
}

Nepal_cluster_freqs[[6]] <- rep(NA, no_Nepal_PP)
# 901 data points in total
Nepal_cluster_freqs_preVac <- rep(0, no_Nepal_PP)
for (i in 1:length(Nepal_time_points_preVac)) {
  Nepal_cluster_freqs_preVac <- Nepal_cluster_freqs_preVac + replace(Nepal_cluster_freqs[[i]],which(is.na(Nepal_cluster_freqs[[i]])),0)
}

for (i in 1:length(Nepal_time_points_selected)) {
  file_name <- paste("Nepal_cluster_freqs_", as.character(i),".rds", sep = "")
  saveRDS(Nepal_cluster_freqs[[i]], file_name)
}
```

#create initial population
```{r}
### create initial population that is based on the first time point data set
# sample from it with an Poisson process
expand_factor <- 15000 / sum(Nepal_cluster_freqs_preVac)
exp_noise <- 10
Nepal_model_start_pop <- (sapply((Nepal_cluster_freqs_preVac + rexp(n = length(Nepal_cluster_freqs_preVac), rate = exp_noise)) * expand_factor, rpois, n=1))

# visual check for the "sampling"
plot(Nepal_cluster_freqs_preVac/sum(Nepal_cluster_freqs_preVac))
points(Nepal_model_start_pop/sum(Nepal_model_start_pop), col = "red")

# aha, I compared the data but 90% of the data are actually coming from 2009 (of the preVac era)
# so it would be much better to only use that time point I think

Nepal_model_start_pop <- (sapply((Nepal_cluster_freqs[[5]] + rexp(n = length(Nepal_cluster_freqs[[5]]), rate = exp_noise)) * expand_factor, rpois, n=1))

# visual check for the "sampling"
plot(Nepal_cluster_freqs[[5]]/sum(Nepal_cluster_freqs[[5]]))
points(Nepal_model_start_pop/sum(Nepal_model_start_pop), col = "red")

saveRDS(Nepal_model_start_pop, "Nepal_model_start_pop.rds")
```


# calculate delta ranking
# I will do last data point - all pre_vac data points combined
```{r}
# For strong and weak selection in the model, I need to calculate the beta statistics.

# calculate gene frequencies first, separate for three time points
Nepal_gene_freq_preVac <- apply(Nepal_ggCaller_intermed_preVac[-1,-1], 1, sum_as_int)
Nepal_gene_freq_postVac <- apply(Nepal_ggCaller_byYear_intermed[[length(Nepal_ggCaller_byYear_intermed)]][-1,-1], 1, sum_as_int)


# first, calculate pre/peri and post vacc frequencies of genes:
Nepal_gene_freq_preVac_rel <- Nepal_gene_freq_preVac/sum(Nepal_gene_freq_preVac)
Nepal_gene_freq_postVac_rel <- Nepal_gene_freq_postVac/sum(Nepal_gene_freq_postVac)

# calculate delta statistic (refer to Corander et al. for more info)
Nepal_delta_data <- (Nepal_gene_freq_postVac_rel - Nepal_gene_freq_preVac_rel) ^ 2 / (1 - Nepal_gene_freq_preVac_rel * (1 - Nepal_gene_freq_preVac_rel))
Nepal_delta_ranking <- rank(Nepal_delta_data)

saveRDS(Nepal_delta_ranking,"Nepal_delta_ranking.rds")
```

