---
title: "WrightFisher_Massachusetts_616compartments"
output: html_document
date: "2023-08-07"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
# reading in the massachussets preprocessed data we received from Nick Croucher
# it contains info about serotype, vaccine type (VT), sequence cluster (SC) and the gene-presence-absence matrix
mass_data <- read.delim("~/Documents/PhD_Project/Data/Massachusetts_data_NickCroucher/mass.input")

```

```{r}
# compute gene frequencies
sum_as_int <- function(x){
  sum(as.integer(x))
}

gene_freq <- rep(0, ncol(mass_data)-5)
gene_freq_mass_data <- mass_data[1:nrow(mass_data),6:ncol(mass_data)]

gene_freq <- apply(gene_freq_mass_data,2, sum_as_int)
gene_freq <- gene_freq / (length(gene_freq_mass_data[,1]))

#> sum(gene_freq <= 0.95 & gene_freq >= 0.05)
#[1] 1062
#> sum(gene_freq <= 1)
#[1] 5425

#create sub-data set for the intermediate frequency genes (5-95% across all sequences)
intermed_mass_data <- data.frame(matrix(ncol = sum(gene_freq <= 0.95 & gene_freq >= 0.05) + 5, nrow = nrow(mass_data)))
intermed_mass_data[,1:5] <- mass_data[,1:5]
colnames(intermed_mass_data)[1:5] <- colnames(mass_data)[1:5]
counter <- 1
for (i in 1:length(gene_freq)){
  if (0.05 <= gene_freq[i] & gene_freq[i] <= 0.95 ){
    intermed_mass_data[,counter+5] <- mass_data[,i+5]
    colnames(intermed_mass_data)[counter+5] <- colnames(mass_data)[i+5]
    counter <- counter + 1
  }
}
```

```{r}
mass_clusters <- length(unique(mass_data$SC))
#calculate the frequency of the sequence clusters and year
mass_cluster_freq_1 <- rep(0,mass_clusters)
mass_cluster_freq_2 <- rep(0,mass_clusters)
mass_cluster_freq_3 <- rep(0,mass_clusters)
for (i in 1:mass_clusters){
  mass_cluster_freq_1[i] <- length(which(mass_data[which(mass_data$SC==i-1),]$Time==0))
  mass_cluster_freq_2[i] <- length(which(mass_data[which(mass_data$SC==i-1),]$Time==36))
  mass_cluster_freq_3[i] <- length(which(mass_data[which(mass_data$SC==i-1),]$Time==72))
}

```

```{r}
# I want to include low and high levels of selection to the model.
# for that I need to calculate the beta statistics

# calculate gene frequencies first, separate for three times
#mass_gene_freq_0 <- rep(0, ncol(mass_data)-5)
#mass_gene_freq_36 <- rep(0, ncol(mass_data)-5)
#mass_gene_freq_72 <- rep(0, ncol(mass_data)-5)

mass_gene_freq_0 <- apply(intermed_mass_data[which(intermed_mass_data$Time==0),][,-(1:5)], 2, sum)
mass_gene_freq_36 <- apply(intermed_mass_data[which(intermed_mass_data$Time==36),][,-(1:5)], 2, sum)
mass_gene_freq_72 <- apply(intermed_mass_data[which(intermed_mass_data$Time==72),][,-(1:5)], 2, sum)

# first, calculate pre/peri and post vacc frequencies of genes:
pre_peri_vacc_gene_freq <- (mass_gene_freq_0 + mass_gene_freq_36) / (nrow(subset(mass_data,mass_data$Time==0)) + nrow(subset(mass_data,mass_data$Time==36)))
#pre_vacc_gene_freq <- (data_gene_freq_1) / (nrow(subset(clusters,clusters$seqYear==2001)))
post_vacc_gene_freq <- mass_gene_freq_72 / nrow(subset(mass_data,mass_data$Time==72))

# calculate delta statistic (refer to Corander et al. for more info)
delta_data <- (post_vacc_gene_freq - pre_peri_vacc_gene_freq) ^ 2 / (1 - pre_peri_vacc_gene_freq * (1 - pre_peri_vacc_gene_freq))
delta_ranking <- rank(delta_data)

```

```{r}
### create initial population that is based on the 2001 data set but not an exact sampling from it
# but a Poisson process
expand_factor <- 15000 / sum(mass_cluster_freq_1)
exp_noise <- 1e2
start_vector <- as.integer(mass_data$Time==0)
  #append(rep(1,sum(mass_cluster_freq_1)), rep(0,(sum(mass_cluster_freq_2)+sum(mass_cluster_freq_3))))
model_start_pop_616 <- (sapply((start_vector + rexp(n = length(start_vector), rate = exp_noise)) * expand_factor, rpois, n=1))

model_start_pop_616clusters <- rep(0, mass_clusters)
for (i in 1:mass_clusters){
  model_start_pop_616clusters[i] <- sum(model_start_pop_616[which(mass_data$SC==i-1)])
}
# visual check for the "sampling"
par(mfrow=c(1,1))
op <- par(mar = c(5,7,4,2) + 0.1,mgp=c(3,1,0))
plot(1:mass_clusters, mass_cluster_freq_1/sum(mass_cluster_freq_1))
points(model_start_pop_616clusters/sum(model_start_pop_616clusters), col = "red")
# looks similar enough, I would say
```

### READ IN THE MODEL FROM THE MODEL FILE
```{r}
# install.packages("drat") # -- if you don't have drat installed
# drat:::add("ncov-ic")
# install.packages("odin.dust")
library(odin.dust)
```

```{r}
WF_nG_h_vP <- odin.dust::odin_dust("WrightFisher_newData_nGenotypes_haploid_PopsizeVariablePois.R")
```

#define parameters for the model:
# watch out: clusters in this version are now single sequences
```{r}
seq_no <- nrow(mass_data)
avg_cluster_freq <- rep(1/seq_no, seq_no)

genotype_df <- t(intermed_mass_data)
genotype_df <- genotype_df[-(1:5),]
genotype_df <- apply(genotype_df,1, as.double)
genotype_df <- t(genotype_df)

params_n_vP <- list(dt = 1/36, species_no = seq_no,  gene_no = nrow(genotype_df), Pop_ini = as.double(model_start_pop_616), Pop_eq = as.double(model_start_pop_616), capacity = sum(model_start_pop_616), Genotypes = as.matrix(genotype_df), sigma_f = 0.1363, sigma_w = 0.0023, prop_f = 0.2483, delta = delta_ranking, m = 0.0044, migVec = avg_cluster_freq, vaccTypes = as.double(mass_data$VT), v = 0.1, vacc_time = 0)
```

### Running the model:
```{r}
WFmodel_nG_h_vP <- WF_nG_h_vP$new(pars = params_n_vP,
                         time = 1,
                         n_particles = 10L,
                         n_threads = 4L,
                         seed = 1L)
# this is expecting doubles. I need to work on that.
```

```{r fig.height=5, fig.width=7}
par(mfrow=c(1,1))
# dt <- 1
dt <- 1/36
n_particles <- 10L
#WFmodel_nG_h_vP$update_state(pars = params_n_vP, time = 0)
WFmodel_nG_h_vP <- WF_nG_h_vP$new(pars = params_n_vP,
                         time = 1,
                         n_particles = 10L,
                         n_threads = 4L,
                         seed = 1L)


n_times <- 180
n_times <- 73
x <- array(NA, dim = c(WFmodel_nG_h_vP$info()$len, n_particles, n_times))

for (t in seq_len(n_times)) {
  x[ , , t] <- WFmodel_nG_h_vP$run(t)
}
time <- x[1, 1, ]
x <- x[-1, , ]



par(mar = c(4.1, 5.1, 0.5, 0.5), las = 1)
if (length(x[,1,1]) <= 8){
  cols <- c("#000000", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7") #8 colorblind friendly colors
} else {
  cols <- rainbow(length(x[,1,1]))
}


library(dplyr)
library(zoo)
par(ask = FALSE)
matplot(time, rollmean(x[1,1 , ], k=36, fill=NA), type = "l",
         xlab = "Time", ylab = "Number of individuals",
         col = cols[1], lty = 1, ylim = range(x))
for (species in 2:length(x[,1,1])) {
  matlines(time, rollmean(x[species,1 , ], k=36, fill=NA), col = cols[species], lty = 1)
}
# or 
 matplot(time,(x[1,1 , ]), type = "l",
         xlab = "Time", ylab = "Number of individuals",
         col = cols[1], lty = 1, ylim = range(x))
for (species in 2:length(x[,1,1])) {
  matlines(time, (x[species,1 , ]), col = cols[species], lty = 1)
}
#if you want to see all single values, not the monthly rolling mean


legend("bottomright", lwd = 1, col = cols[1:length(x[,1,1])], legend = 1:length(x[,1,1]), bty = "n")
```

# create function for creating lollipop plots
```{r}


library(ggplot2)
library(gridExtra)

make_lollipop_plot <- function(data1, data2, data3, model1, model2, model3){
  
# Create data
          lollipop_data_2001_nVT <- data.frame(
            x=which(mass_VT==0),
            model2001=model1[which(mass_VT==0)] / sum(model1),
            data2001=as.numeric(data1[which(mass_VT==0)] / sum(data1))
          )
          lollipop_data_2001_VT <- data.frame(
            x=which(mass_VT==1),
            model2001=model1[which(mass_VT==1)] / sum(model1),
            data2001=as.numeric(data1[which(mass_VT==1)] / sum(data1))
          )
          lollipop_data_2004_nVT <- data.frame(
            x=which(mass_VT==0),
            model2004=model2[which(mass_VT==0)] / sum(model2),
            data2004=as.numeric(data2[which(mass_VT==0)] / sum(data2))
          )
          lollipop_data_2004_VT <- data.frame(
            x=which(mass_VT==1),
            model2004=model2[which(mass_VT==1)] / sum(model2),
            data2004=as.numeric(data2[which(mass_VT==1)] / sum(data2))
          )
          lollipop_data_2007_nVT <- data.frame(
            x=which(mass_VT==0),
            model2007=model3[which(mass_VT==0)] / sum(model3),
            data2007=as.numeric(data3[which(mass_VT==0)] / sum(data3))
          )
          lollipop_data_2007_VT <- data.frame(
            x=which(mass_VT==1),
            model2007=model3[which(mass_VT==1)] / sum(model3),
            data2007=as.numeric(data3[which(mass_VT==1)] / sum(data3))
          )
          # Change baseline
          lollipop_plot_2001_nVT <- ggplot(lollipop_data_2001_nVT) +
            geom_segment( aes(x=which(mass_VT==0), xend=which(mass_VT==0), y=model2001, yend=data2001), color="grey") +
            geom_point( aes(x=which(mass_VT==0), y=model2001, color="Model non-VT"), size=3 ) +
            geom_point( aes(x=which(mass_VT==0), y=data2001, color="Data non_VT"), size=3 ) +
            scale_color_manual(values = c("#E69F00", "#56B4E9"),
                               guide  = guide_legend(), 
                               name   = "Group") +
            coord_flip()+
            #theme_ipsum() +
            theme(legend.position = "none") +
            ggtitle("2001 Non-Vaccine Types") +
            ylab("Frequency") +
            xlab("Clusters") +
            theme(axis.title  = element_text(size = 20), axis.text = element_text(size = 20), plot.title = element_text(size = 25,hjust = 0.5))  +
            ylim(0, max(max(lollipop_data_2001_nVT$model2001),max(lollipop_data_2001_nVT$data2001),max(lollipop_data_2004_VT$model2004),max(lollipop_data_2004_VT$data2004),max(lollipop_data_2007_nVT$model2007),max(lollipop_data_2007_nVT$data2007)))
          lollipop_plot_2001_VT <- ggplot(lollipop_data_2001_VT) +
            geom_segment( aes(x=which(mass_VT==1), xend=which(mass_VT==1), y=model2001, yend=data2001), color="grey") +
            geom_point( aes(x=which(mass_VT==1), y=model2001, color="Model non-VT"), size=3 ) +
            geom_point( aes(x=which(mass_VT==1), y=data2001, color="Data non_VT"), size=3 ) +
            scale_color_manual(values = c("#E69F00", "#56B4E9"),
                               guide  = guide_legend(), 
                               name   = "Group") +
            coord_flip()+
            #theme_ipsum() +
            theme(legend.position = "none") +
            ggtitle("2001 Vaccine Types") +
            ylab("Frequency") +
            xlab("Clusters") +
            theme(axis.title  = element_text(size = 20), axis.text = element_text(size = 20), plot.title = element_text(size = 25,hjust = 0.5))  +
            ylim(0, max(max(lollipop_data_2001_nVT$model2001),max(lollipop_data_2001_nVT$data2001),max(lollipop_data_2004_VT$model2004),max(lollipop_data_2004_VT$data2004),max(lollipop_data_2007_nVT$model2007),max(lollipop_data_2007_nVT$data2007)))
          lollipop_plot_2004_nVT <- ggplot(lollipop_data_2004_nVT) +
            geom_segment( aes(x=which(mass_VT==0), xend=which(mass_VT==0), y=model2004, yend=data2004), color="grey") +
            geom_point( aes(x=which(mass_VT==0), y=model2004, color="Model"), size=3 ) +
            geom_point( aes(x=which(mass_VT==0), y=data2004, color="Data"), size=3 ) +
            scale_color_manual(values = c("#E69F00", "#56B4E9"),
                               guide  = guide_legend(), 
                               name   = "Group") +
            coord_flip()+
            #theme_ipsum() +
            theme(legend.position = "none") +
            ggtitle("2004 Non-Vaccine Types") +
            ylab("Frequency") +
            xlab("Clusters") +
            theme(axis.title  = element_text(size = 20), axis.text = element_text(size = 20), plot.title = element_text(size = 25,hjust = 0.5))  +
ylim(0, max(max(lollipop_data_2001_nVT$model2001),max(lollipop_data_2001_nVT$data2001),max(lollipop_data_2004_VT$model2004),max(lollipop_data_2004_VT$data2004),max(lollipop_data_2007_nVT$model2007),max(lollipop_data_2007_nVT$data2007)))          
          lollipop_plot_2004_VT <- ggplot(lollipop_data_2004_VT) +
            geom_segment( aes(x=which(mass_VT==1), xend=which(mass_VT==1), y=model2004, yend=data2004), color="grey") +
            geom_point( aes(x=which(mass_VT==1), y=model2004, color="Model"), size=3 ) +
            geom_point( aes(x=which(mass_VT==1), y=data2004, color="Data"), size=3 ) +
            scale_color_manual(values = c("#E69F00", "#56B4E9"),
                               guide  = guide_legend(), 
                               name   = "Group") +
            coord_flip()+
            #theme_ipsum() +
            theme(legend.position = "none") +
            ggtitle("2004 Vaccine Types") +
            ylab("Frequency") +
            xlab("Clusters") +
            theme(axis.title  = element_text(size = 20), axis.text = element_text(size = 20), plot.title = element_text(size = 25,hjust = 0.5))  +
            ylim(0, max(max(lollipop_data_2001_nVT$model2001),max(lollipop_data_2001_nVT$data2001),max(lollipop_data_2004_VT$model2004),max(lollipop_data_2004_VT$data2004),max(lollipop_data_2007_nVT$model2007),max(lollipop_data_2007_nVT$data2007))) 
          lollipop_plot_2007_nVT <- ggplot(lollipop_data_2007_nVT) +
            geom_segment( aes(x=which(mass_VT==0), xend=which(mass_VT==0), y=model2007, yend=data2007), color="grey") +
            geom_point( aes(x=which(mass_VT==0), y=model2007, color="Model"), size=3 ) +
            geom_point( aes(x=which(mass_VT==0), y=data2007, color="Data"), size=3 ) +
            scale_color_manual(values = c("#E69F00", "#56B4E9"),
                               guide  = guide_legend(), 
                               name   = "Group") +
            coord_flip()+
            #theme_ipsum() +
            theme(legend.position = "none") +
            ggtitle("2007 Non-Vaccine Types") +
            ylab("Frequency") +
            xlab("Clusters") +
            theme(axis.title  = element_text(size = 20), axis.text = element_text(size = 20), plot.title = element_text(size = 25,hjust = 0.5))  +
            ylim(0, max(max(lollipop_data_2001_nVT$model2001),max(lollipop_data_2001_nVT$data2001),max(lollipop_data_2004_VT$model2004),max(lollipop_data_2004_VT$data2004),max(lollipop_data_2007_nVT$model2007),max(lollipop_data_2007_nVT$data2007)))
          lollipop_plot_2007_VT <- ggplot(lollipop_data_2007_VT) +
            geom_segment( aes(x=which(mass_VT==1), xend=which(mass_VT==1), y=model2007, yend=data2007), color="grey") +
            geom_point( aes(x=which(mass_VT==1), y=model2007, color="Model"), size=3 ) +
            geom_point( aes(x=which(mass_VT==1), y=data2007, color="Data"), size=3 ) +
            scale_color_manual(values = c("#E69F00", "#56B4E9"),
                               guide  = guide_legend(), 
                               name   = "Legend") +
            coord_flip()+
            #theme_ipsum() +
            theme(legend.position = c(.8,.8),legend.text = element_text(size = 20),legend.title = element_text(size = 20)) +
            ggtitle("2007 Vaccine Types") +
            ylab("Frequency") +
            xlab("Clusters") +
            theme(axis.title  = element_text(size = 20), axis.text = element_text(size = 20), plot.title = element_text(size = 25,hjust = 0.5))  +
            ylim(0, max(max(lollipop_data_2001_nVT$model2001),max(lollipop_data_2001_nVT$data2001),max(lollipop_data_2004_VT$model2004),max(lollipop_data_2004_VT$data2004),max(lollipop_data_2007_nVT$model2007),max(lollipop_data_2007_nVT$data2007)))
          
          grid.arrange(lollipop_plot_2001_nVT + theme(plot.margin = unit(c(.5,0.5,1,0.5), "cm")),lollipop_plot_2004_nVT+ theme(plot.margin = unit(c(.5,0.5,1,0.5), "cm")), lollipop_plot_2007_nVT + theme(plot.margin = unit(c(.5,0.5,1,0.5), "cm")),lollipop_plot_2001_VT + theme(plot.margin = unit(c(.5,0.5,.5,0.5), "cm")), lollipop_plot_2004_VT + theme(plot.margin = unit(c(.5,0.5,.5,0.5), "cm")), lollipop_plot_2007_VT + theme(plot.margin = unit(c(.5,0.5,.5,0.5), "cm")), ncol = 3, nrow=2)
}
```

# create lollipop plot for testing different parameter values manually
```{r}
# calculate Vaccine Type consensus for clusters
mass_VT <- rep(0, mass_clusters)
mass_VT_mean <- rep(0, mass_clusters)

for (i in 1:mass_clusters){
  curr_VTs <- subset(mass_data,mass_data$SC == i-1)$VT # select all genomes in cluster i-1
  mass_VT[i] <- ceiling(median(curr_VTs))
  mass_VT_mean[i] <- mean(curr_VTs)
  #print(mass_VT_mean[i]) #calculate mean as check for consensus
}
model_616clusters_1 <- rep(0, mass_clusters)
model_616clusters_37 <- rep(0, mass_clusters)
model_616clusters_73 <- rep(0, mass_clusters)
for (i in 1:mass_clusters){
  model_616clusters_1[i] <- sum(x[which(mass_data$SC==i-1),1,1])
  model_616clusters_37[i] <- sum(x[which(mass_data$SC==i-1),1,37])
  model_616clusters_73[i] <- sum(x[which(mass_data$SC==i-1),1,73])
}
make_lollipop_plot(data1 = mass_cluster_freq_1, data2 = mass_cluster_freq_2, data3= mass_cluster_freq_3, model1 = model_616clusters_1, model2 = model_616clusters_37, model3 = model_616clusters_73)
```


### now, the fitting can begin! :-)
```{r}
#install.packages("drat") # -- if you don't have drat installed
drat:::add("ncov-ic")
#install.packages("mcstate")
library(mcstate)
```

```{r}
# process data with particle filter:
dt <- 1/36 # we assume that the generation time of Strep. pneumo is 1 month
# we have data from 2001, 2004, 2007, so we want 3 (years) * 12 (months) = 36 updates in-between

peripost_mass_cluster_freq <- data.frame("year" = c(1, 2), rbind(mass_cluster_freq_2, mass_cluster_freq_3))
names(peripost_mass_cluster_freq) <- c("year", as.character(1:mass_clusters))
  

# (until now I had 2004 and 2007 there and then in the model that resulted in times 72144 and 72252. does not make sense)
# when changing this back, also change initial_time back to 2001

fitting_mass_data <- mcstate::particle_filter_data(data = peripost_mass_cluster_freq,
                                          time = "year",
                                          rate = 1 / dt,
                                          initial_time = 0)
```

### Likelihood:
```{r}

# log-likelihood of with Poisson distribution
ll_pois <- function(obs, model) {
  exp_noise <- 1e6

  if (is.na(obs)) {
    # Creates vector of zeros in ll with same length, if no data
    ll_obs <- numeric(length(model))
  } else {
    lambda <- model + rexp(n = length(model), rate = exp_noise)
    ll_obs <- dpois(x = obs, lambda = lambda, log = TRUE)
    #print(model)
  }
  ll_obs
}

combined_compare_sum <- function(state, observed, pars = NULL) {
  result <- 0
  #data_size <- sum(observed[as.character(1:62)])
  # cannot get this to work for now
  # instead I will use this:
  data_size <- sum(mass_cluster_freq_1)
  #model_size <- sum(state[, , drop = TRUE])
  model_size = 15000
  
  for (i in 1:mass_clusters){
    #browser()
    #print(observed[[as.character(i)]])
    #print(state[1+i, , drop = TRUE]/model_size * data_size)
    #print(state[1+i, , drop = TRUE])
    state_cluster <- sum(state[which(mass_data$SC==i-1),,drop = TRUE])
    result <- result + ll_pois(observed[[as.character(i)]], state_cluster/model_size * data_size)
    #test_likelihood <- ll_pois(observed[[as.character(i)]], state[1+i, , drop = TRUE]/model_size * data_size)
    #print(result)
  }
  #print(result)
  result
}

#manual usage: combined_compare(WFmodel_nG_h_vP$state(), observed = mass_data[1,])
```

```{r}
#index <- function(info) {
#  list(run = 5L, state = 2:63)
#}
### trying out deterministic particle filter
det_filter <- particle_deterministic$new(data = fitting_mass_data,
                                       model = WF_nG_h_vP,
                                       compare = combined_compare_sum)
```

```{r}
# run determinstic filter
params_n_vP <- list(dt = 1/36, species_no = seq_no,  gene_no = nrow(genotype_df), Pop_ini = as.double(model_start_pop_616), Pop_eq = as.double(model_start_pop_616), capacity = sum(model_start_pop_616), Genotypes = as.matrix(genotype_df), sigma_f = 0.1363, sigma_w = 0.0023, prop_f = 0.2483, delta = delta_ranking, m = 0.0044, migVec = avg_cluster_freq, vaccTypes = as.double(mass_data$VT), v = 0.1, vacc_time = 0)
testfilter <- det_filter$run(save_history = TRUE, pars = params_n_vP)
```

#Plot particle filter
```{r}
plot_particle_filter_sum <- function(history, true_history, times, obs_end = NULL) {
  if (is.null(obs_end)) {
    obs_end <- max(times)
  }
  #summarise over compartments in model to create clusters
  history_clusters <- matrix(dat = NA, nrow = mass_clusters, ncol = length(times))
  for (i in 1:mass_clusters){
    for (j in times) {
      history_clusters[i,j] <- sum(history[which(mass_data$SC==i-1),,j])
    }
  }
  
  par(mar = c(4.1, 5.1, 0.5, 0.5), las = 1)
  cols <- rainbow(mass_clusters)
  
  matplot(times, history_clusters[1,], type = "l",
          xlab = "Time", ylab = "Number of individuals",
          col = cols[1], lty = 1, ylim = range(history_clusters[1:mass_clusters,]))
  for (species in 2:length(history_clusters[,1])) {
    matlines(times, (history_clusters[species,]), col = cols[species-1], lty = 1)
  }  
  for (species in 2:mass_clusters) {
    matpoints(times[1:obs_end], (true_history[1:2,species]), pch = 19, col = cols[species-1])
  }
  legend("left", lwd = 1, col = cols, legend = 1:62, bty = "n")
  
}

scaled_filter_history <- det_filter$history()/15000 * sum(mass_cluster_freq_1)
plot_particle_filter_sum(scaled_filter_history, peripost_mass_cluster_freq,1:2)
```

```{r}
# Using MCMC to infer parameters
#pmcmc_sigma_f <- mcstate::pmcmc_parameter("sigma_f", 0.15, min = 0)
#pmcmc_sigma_w <- mcstate::pmcmc_parameter("sigma_w", 0.05, min = 0)
pmcmc_sigma_f <- mcstate::pmcmc_parameter("sigma_f", 0.15, min = 0, max = 1)
pmcmc_sigma_w <- mcstate::pmcmc_parameter("sigma_w", 0.05, min = 0, max = 1)
pmcmc_prop_f <- mcstate::pmcmc_parameter("prop_f", 0.25, min = 0, max = 1)
pmcmc_m <- mcstate::pmcmc_parameter("m", 0.03, min = 0, max = 1)
pmcmc_v <- mcstate::pmcmc_parameter("v", 0.05, min = 0, max = 1)
species_no <- seq_no
no_clusters <- seq_no
gene_no <- nrow(as.matrix(genotype_df))
#Pop_ini <- cluster_freq_1
#Pop_eq <- cluster_freq_1
Pop_ini <- model_start_pop_616
Pop_eq <- model_start_pop_616
#Genotypes <- as.matrix(mass_consensus_presence_absence[,-1])
Genotypes <- as.matrix(genotype_df)

#capacity <- sum(cluster_freq_1)
capacity <- sum(model_start_pop_616)
delta <- delta_ranking
vaccTypes <- as.double(mass_data$VT)
vacc_time <- 0
dt <- 1/36
migVec <- avg_cluster_freq

complex_params <- c(Pop_ini, Pop_eq, Genotypes, capacity, delta, vaccTypes, species_no, gene_no, vacc_time, dt, migVec)
#set_names <- function(x, nms) {
#  names(x) <- nms
#  x
#}
#transform <- set_names(lapply(complex_params, make_transform), complex_params)
make_transform <- function(p) {
  function(theta){
        c(list(Pop_ini = p[1:species_no],
         Pop_eq = p[(species_no +1) : (species_no + species_no)],
         Genotypes = matrix(p[(species_no + species_no + 1): ((species_no + species_no + 1) + (gene_no * species_no) - 1)], nrow = gene_no, ncol = species_no),
         capacity = p[((2 * species_no + 1) + (gene_no * species_no) - 1) + 1],
         delta = p[(((2 * species_no + 1) + (gene_no * species_no) - 1) + 2) : (((2 * species_no + 1) + (gene_no * species_no) - 1) + 2 + gene_no -1)],
         vaccTypes = p[(((2 * species_no + 1) + (gene_no * species_no) - 1) + 2 + gene_no) : ((((2 * species_no + 1) + (gene_no * species_no) - 1) + 2 + gene_no) + no_clusters -1)],
         species_no = p[(((2 * species_no + 1) + (gene_no * species_no) - 1) + 2 + gene_no) + no_clusters],
         gene_no = p[(((2 * species_no + 1) + (gene_no * species_no) - 1) + 2 + gene_no) + no_clusters + 1],
         vacc_time = p[(((2 * species_no + 1) + (gene_no * species_no) - 1) + 2 + gene_no) + no_clusters + 2],
         dt = p[(((2 * species_no + 1) + (gene_no * species_no) - 1) + 2 + gene_no) + no_clusters + 3],
         migVec = p[((((2 * species_no + 1) + (gene_no * species_no) - 1) + 2 + gene_no) + no_clusters + 4):((((2 * species_no + 1) + (gene_no * species_no) - 1) + 2 + gene_no) + no_clusters + 4 + species_no - 1)]), as.list(theta))
  }

  }

transform <- function(x) {
  make_transform(complex_params)}
proposal_matrix <- diag(0.1, 5) # the proposal matrix defines the covariance-variance matrix for a mult normal dist
# here, all parameters are proposed independently.
# think about this, this might not actually be true (should be okay as a first assessment though)
# but maybe I want larger variance for some of the parameters
proposal_matrix <- diag(x = c(0.3,0.1,0.3,0.3,0.3), 5)
#mcmc_pars <- mcstate::pmcmc_parameters$new(list(pmcmc_sigma_f, pmcmc_sigma_w, pmcmc_prop_f, pmcmc_m, pmcmc_v), proposal_matrix, transform)
mcmc_pars <- mcstate::pmcmc_parameters$new(list(mcstate::pmcmc_parameter("sigma_f", 0.15, min = 0.075, max = 1), mcstate::pmcmc_parameter("sigma_w", 0.05, min = 0, max = 0.0749), mcstate::pmcmc_parameter("prop_f", 0.25, min = 0, max = 1), mcstate::pmcmc_parameter("m", 0.03, min = 0, max = 0.5), mcstate::pmcmc_parameter("v", 0.05, min = 0, max = 0.5)), proposal_matrix, make_transform(complex_params))
#= make_transform(c(Pop_ini, Pop_eq, Genotypes, capacity, delta, vaccTypes, species_no, gene_no, vacc_time)))
#mcmc_pars$names()
#mcmc_pars$model(mcmc_pars$initial())
mcmc_pars$initial()
# read this: https://mrc-ide.github.io/mcstate/reference/pmcmc_parameters.html
# it explains how to not fit all parameters but just the ones I want
# non-scalar parameters have to be transformed for this.
```

### DETERMINISTIC FITTING

```{r}
# do fitting with deterministic particle filter
#n_particles <- 20
det_filter <- particle_deterministic$new(data = fitting_mass_data,
                                       model = WF_nG_h_vP,
                                       compare = combined_compare_sum)

n_steps <- 500
n_burnin <- 0


control <- mcstate::pmcmc_control(
    n_steps,
    save_state = TRUE, 
    save_trajectories = TRUE,
    progress = TRUE,
    adaptive_proposal = TRUE,
    n_chains = 1)
det_pmcmc_run <- mcstate::pmcmc(mcmc_pars, det_filter, control = control)
par(mfrow = c(1,1))

#plot_particle_filter(det_pmcmc_run$trajectories$state/15000*300, peripost_mass_cluster_freq, 1:2)
library(coda)
det_mcmc1 <- coda::as.mcmc(cbind(det_pmcmc_run$probabilities, det_pmcmc_run$pars))
plot(det_mcmc1)


```


```{r}
#use lollipop plotting function
det_pmcmc_run_no_time <- det_pmcmc_run$trajectories$state[-1,,]
modelfit_616clusters_1 <- rep(0, mass_clusters)
modelfit_616clusters_2 <- rep(0, mass_clusters)
modelfit_616clusters_3 <- rep(0, mass_clusters)
for (i in 1:mass_clusters){
  modelfit_616clusters_1[i] <- sum(det_pmcmc_run_no_time[which(mass_data$SC==i-1),1,1])
  modelfit_616clusters_2[i] <- sum(det_pmcmc_run_no_time[which(mass_data$SC==i-1),1,2])
  modelfit_616clusters_3[i] <- sum(det_pmcmc_run_no_time[which(mass_data$SC==i-1),1,3])
}

make_lollipop_plot(data1 = mass_cluster_freq_1, data2 = mass_cluster_freq_2, data3 = mass_cluster_freq_3, model1 = modelfit_616clusters_1, model2 = modelfit_616clusters_2, model3 = modelfit_616clusters_3)

```

```{r}
det_proposal_matrix <- cov(det_pmcmc_run$pars)
det_mcmc_pars <- mcstate::pmcmc_parameters$new(list(mcstate::pmcmc_parameter("sigma_f", 0.15, min = 0.075, max = 0.5), mcstate::pmcmc_parameter("sigma_w", 0.005, min = 0, max = 0.05), mcstate::pmcmc_parameter("prop_f", 0.25, min = 0, max = 1), mcstate::pmcmc_parameter("m", 0.03, min = 0, max = 0.1), mcstate::pmcmc_parameter("v", 0.05, min = 0, max = 1)), det_proposal_matrix, make_transform(complex_params))
det_proposal_matrix
```

```{r}
det_filter <- particle_deterministic$new(data = fitting_mass_data,
                                       model = WF_nG_h_vP,
                                       compare = combined_compare_sum)

n_steps <- 1500
n_burnin <- 0


control <- mcstate::pmcmc_control(
    n_steps,
    save_state = TRUE, 
    save_trajectories = TRUE,
    progress = TRUE,
    adaptive_proposal = TRUE,
    n_chains = 1)
det_pmcmc_run2 <- mcstate::pmcmc(det_mcmc_pars, det_filter, control = control)
par(mfrow = c(1,1))

#plot_particle_filter(det_pmcmc_run$trajectories$state/15000*300, peripost_mass_cluster_freq, 1:2)
library(coda)
det_mcmc2 <- coda::as.mcmc(cbind(det_pmcmc_run2$probabilities, det_pmcmc_run2$pars))
plot(det_mcmc2)
```

```{r}
processed_chains <- mcstate::pmcmc_thin(det_pmcmc_run2, burnin = 500, thin = 1)
parameter_mean_hpd <- apply(processed_chains$pars, 2, mean)
parameter_mean_hpd
```

```{r}
#use lollipop plotting function
det_pmcmc_run_no_time <- det_pmcmc_run2$trajectories$state[-1,,]
modelfit_616clusters_1 <- rep(0, mass_clusters)
modelfit_616clusters_2 <- rep(0, mass_clusters)
modelfit_616clusters_3 <- rep(0, mass_clusters)
for (i in 1:mass_clusters){
  modelfit_616clusters_1[i] <- sum(det_pmcmc_run_no_time[which(mass_data$SC==i-1),1,1])
  modelfit_616clusters_2[i] <- sum(det_pmcmc_run_no_time[which(mass_data$SC==i-1),1,2])
  modelfit_616clusters_3[i] <- sum(det_pmcmc_run_no_time[which(mass_data$SC==i-1),1,3])
}

make_lollipop_plot(data1 = mass_cluster_freq_1, data2 = mass_cluster_freq_2, data3 = mass_cluster_freq_3, model1 = modelfit_616clusters_1, model2 = modelfit_616clusters_2, model3 = modelfit_616clusters_3)

```

```{r}
det_proposal_matrix <- cov(det_pmcmc_run2$pars)
det_mcmc_pars <- mcstate::pmcmc_parameters$new(list(mcstate::pmcmc_parameter("sigma_f", 0.15, min = 0.075, max = 0.5), mcstate::pmcmc_parameter("sigma_w", 0.005, min = 0, max = 0.05), mcstate::pmcmc_parameter("prop_f", 0.25, min = 0, max = 1), mcstate::pmcmc_parameter("m", 0.03, min = 0, max = 0.1), mcstate::pmcmc_parameter("v", 0.05, min = 0, max = 1)), det_proposal_matrix, make_transform(complex_params))
det_proposal_matrix
```

```{r}
det_filter <- particle_deterministic$new(data = fitting_mass_data,
                                       model = WF_nG_h_vP,
                                       compare = combined_compare_sum)

n_steps <- 5000
n_burnin <- 0


control <- mcstate::pmcmc_control(
    n_steps,
    save_state = TRUE, 
    save_trajectories = TRUE,
    progress = TRUE,
    adaptive_proposal = TRUE,
    n_chains = 1)
det_pmcmc_run3 <- mcstate::pmcmc(det_mcmc_pars, det_filter, control = control)
par(mfrow = c(1,1))

#plot_particle_filter(det_pmcmc_run$trajectories$state/15000*300, peripost_mass_cluster_freq, 1:2)
library(coda)
det_mcmc3 <- coda::as.mcmc(cbind(det_pmcmc_run3$probabilities, det_pmcmc_run3$pars))
plot(det_mcmc3)
```

```{r}
processed_chains <- mcstate::pmcmc_thin(det_pmcmc_run3, burnin = 500, thin = 1)
parameter_mean_hpd <- apply(processed_chains$pars, 2, mean)
parameter_mean_hpd
```

```{r}
#use lollipop plotting function
det_pmcmc_run_no_time <- det_pmcmc_run3$trajectories$state[-1,,]
modelfit_616clusters_1 <- rep(0, mass_clusters)
modelfit_616clusters_2 <- rep(0, mass_clusters)
modelfit_616clusters_3 <- rep(0, mass_clusters)
for (i in 1:mass_clusters){
  modelfit_616clusters_1[i] <- sum(det_pmcmc_run_no_time[which(mass_data$SC==i-1),1,1])
  modelfit_616clusters_2[i] <- sum(det_pmcmc_run_no_time[which(mass_data$SC==i-1),1,2])
  modelfit_616clusters_3[i] <- sum(det_pmcmc_run_no_time[which(mass_data$SC==i-1),1,3])
}

make_lollipop_plot(data1 = mass_cluster_freq_1, data2 = mass_cluster_freq_2, data3 = mass_cluster_freq_3, model1 = modelfit_616clusters_1, model2 = modelfit_616clusters_2, model3 = modelfit_616clusters_3)

```